From 1886c9b28f21939ba88fc585228112031f51a3d0 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Sat, 13 Apr 2019 20:32:49 +0300
Subject: [PATCH] Misc changes


diff --git a/api/src/main/java/net/md_5/bungee/api/Callback.java b/api/src/main/java/net/md_5/bungee/api/Callback.java
index 0cccc175..b040daa4 100644
--- a/api/src/main/java/net/md_5/bungee/api/Callback.java
+++ b/api/src/main/java/net/md_5/bungee/api/Callback.java
@@ -6,6 +6,7 @@ package net.md_5.bungee.api;
  *
  * @param <V> the type of result
  */
+@FunctionalInterface // IvanCord - functional
 public interface Callback<V>
 {
 
diff --git a/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java b/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java
index f7459860..bac6b5fb 100644
--- a/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java
+++ b/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java
@@ -1,5 +1,6 @@
 package net.md_5.bungee.api.connection;
 
+import java.util.Collection;
 import java.util.Locale;
 import java.util.Map;
 import java.util.UUID;
@@ -224,6 +225,26 @@ public interface ProxiedPlayer extends Connection, CommandSender
      */
     void chat(String message);
 
+    // IvanCord start - allow chat method with components
+
+    /**
+     * Make this player chat (say something), to the server he is currently on,
+     * message created via the chat component api
+     *
+     * @param message the message to say
+     */
+    void chat(BaseComponent message);
+
+    /**
+     * Make this player chat (say something), to the server he is currently on,
+     * message created via the chat component api
+     *
+     * @param message the message to say
+     */
+    void chat(BaseComponent... message);
+
+    // IvanCord end
+
     /**
      * Get the server which this player will be sent to next time the log in.
      *
@@ -372,4 +393,24 @@ public interface ProxiedPlayer extends Connection, CommandSender
      * @return this player's {@link Scoreboard}
      */
     Scoreboard getScoreboard();
+
+    // IvanCord start - add way to apply potion effects
+    /**
+     * Adds potion effect to this player.
+     * All effect ids can be seen <a href="https://minecraft.gamepedia.com/Status_effect#Data_values">here</a>
+     *
+     * @param effectId id of the effect.
+     * @param amplifier effect amplifier
+     * @param duration effect duration
+     * @param hideParticles should particles be hided or not
+     */
+    void addPotionEffect(int effectId, int amplifier, int duration, boolean hideParticles);
+
+    /**
+     * Gets the potion effects which this player has
+     *
+     * @return potion effects
+     */
+    Collection<Integer> getCurrentPotionEffects();
+    // IvanCord end
 }
diff --git a/log4j/src/main/resources/log4j2.xml b/log4j/src/main/resources/log4j2.xml
index 3b3525f0..1462f89a 100644
--- a/log4j/src/main/resources/log4j2.xml
+++ b/log4j/src/main/resources/log4j2.xml
@@ -3,7 +3,7 @@
     <Appenders>
         <TerminalConsole name="TerminalConsole">
             <PatternLayout>
-                <LoggerNamePatternSelector defaultPattern="%highlightError{[%d{HH:mm:ss} %level] [%logger]: %minecraftFormatting{%msg}%n%xEx}">
+                <LoggerNamePatternSelector defaultPattern="%highlightError{[%d{HH:mm:ss} %level]: [%logger] %minecraftFormatting{%msg}%n%xEx}"> <!-- IvanCord - changed -->
                     <!-- Log root and BungeeCord loggers without prefix -->
                     <PatternMatch key=",BungeeCord" pattern="%highlightError{[%d{HH:mm:ss} %level]: %minecraftFormatting{%msg}%n%xEx}" />
                 </LoggerNamePatternSelector>
@@ -11,7 +11,7 @@
         </TerminalConsole>
         <RollingRandomAccessFile name="File" fileName="logs/latest.log" filePattern="logs/%d{yyyy-MM-dd}-%i.log.gz" immediateFlush="false">
             <PatternLayout>
-                <LoggerNamePatternSelector defaultPattern="[%d{HH:mm:ss}] [%t/%level] [%logger]: %minecraftFormatting{%msg}{strip}%n">
+                <LoggerNamePatternSelector defaultPattern="[%d{HH:mm:ss}] [%t/%level]: [%logger] %minecraftFormatting{%msg}{strip}%n"> <!-- IvanCord - changed -->
                     <!-- Log root and BungeeCord loggers without prefix -->
                     <PatternMatch key=",BungeeCord" pattern="[%d{HH:mm:ss}] [%t/%level]: %minecraftFormatting{%msg}{strip}%n" />
                 </LoggerNamePatternSelector>
diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 15f7c5b5..539f1e03 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -1,9 +1,7 @@
 package net.md_5.bungee;
 
 import com.google.common.base.Preconditions;
-import com.google.common.collect.HashMultimap;
 import com.google.common.collect.ImmutableMap;
-import com.google.common.collect.Multimap;
 import io.netty.bootstrap.Bootstrap;
 import io.netty.channel.Channel;
 import io.netty.channel.ChannelFuture;
@@ -55,6 +53,7 @@ import net.md_5.bungee.protocol.PacketWrapper;
 import net.md_5.bungee.protocol.Protocol;
 import net.md_5.bungee.protocol.packet.Chat;
 import net.md_5.bungee.protocol.packet.ClientSettings;
+import net.md_5.bungee.protocol.packet.EntityEffect;
 import net.md_5.bungee.protocol.packet.Kick;
 import net.md_5.bungee.protocol.packet.PlayerListHeaderFooter;
 import net.md_5.bungee.protocol.packet.PluginMessage;
@@ -129,7 +128,7 @@ public final class UserConnection implements ProxiedPlayer
     private final Collection<UUID> sentBossBars = new HashSet<>();
     // Waterfall start
     @Getter
-    private final Multimap<Integer, Integer> potions = HashMultimap.create();
+    private final Collection<Integer> potions = new HashSet<>(); // IvanCord - handle potions differently
     // Waterfall end
     /*========================================================================*/
     @Getter
@@ -445,6 +444,24 @@ public final class UserConnection implements ProxiedPlayer
         server.getCh().write( new Chat( message ) );
     }
 
+    // IvanCord start
+
+    @Override
+    public void chat(BaseComponent message)
+    {
+        Preconditions.checkState( server != null, "Not connected to server" );
+        server.getCh().write( new Chat( message.toLegacyText() ) );
+    }
+
+    @Override
+    public void chat(BaseComponent... message)
+    {
+        Preconditions.checkState( server != null, "Not connected to server" );
+        server.getCh().write( new Chat( BaseComponent.toLegacyText( message ) ) );
+    }
+
+    // IvanCord end
+
     @Override
     public void sendMessage(String message)
     {
@@ -739,6 +756,22 @@ public final class UserConnection implements ProxiedPlayer
         return serverSentScoreboard;
     }
 
+    // IvanCord start
+
+    @Override
+    public void addPotionEffect(int effectId, int amplifier, int duration, boolean hideParticles)
+    {
+        unsafe().sendPacket( new EntityEffect( getClientEntityId(), effectId, amplifier, duration, hideParticles ) );
+    }
+
+    @Override
+    public Collection<Integer> getCurrentPotionEffects()
+    {
+        return Collections.unmodifiableCollection( potions );
+    }
+
+    // IvanCord end
+
     // Waterfall start
     public boolean isDisableEntityMetadataRewrite() {
         return disableEntityMetadaRewrite;
diff --git a/proxy/src/main/java/net/md_5/bungee/command/CommandEnd.java b/proxy/src/main/java/net/md_5/bungee/command/CommandEnd.java
index d87d0b95..055b3bb6 100644
--- a/proxy/src/main/java/net/md_5/bungee/command/CommandEnd.java
+++ b/proxy/src/main/java/net/md_5/bungee/command/CommandEnd.java
@@ -14,7 +14,7 @@ public class CommandEnd extends Command
 
     public CommandEnd()
     {
-        super( "end", "bungeecord.command.end" );
+        super( "end", "bungeecord.command.end", "stop" ); // IvanCord - alias stop
     }
 
     @Override
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index 1f8a2439..4739050e 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -597,22 +597,30 @@ public class DownstreamBridge extends PacketHandler
         if (this.con.getForgeClientHandler().isForgeUser() && !this.con.getForgeClientHandler().isHandshakeComplete()) {
             throw CancelSendSignal.INSTANCE;
         }
-        con.getPotions().put(rewriteEntityId(entityEffect.getEntityId()), entityEffect.getEffectId());
+        // IvanCord start - rewrite this
+        con.getPotions().add(entityEffect.getEffectId());
+        con.unsafe().sendPacket( entityEffect );
     }
 
     @Override
     public void handle(EntityRemoveEffect removeEffect) throws Exception
     {
         if (con.isDisableEntityMetadataRewrite()) return; // Waterfall
-        con.getPotions().remove(rewriteEntityId(removeEffect.getEntityId()), removeEffect.getEffectId());
+        con.getPotions().remove(removeEffect.getEffectId());
+        con.unsafe().sendPacket( removeEffect );
+        // IvanCord end
     }
 
+    // IvanCord start - this is not needed now
+    /*
     private int rewriteEntityId(int entityId) {
         if (entityId == con.getServerEntityId()) {
             return con.getClientEntityId();
         }
         return entityId;
     }
+     */
+    // IvanCord end
     // Waterfall end
 
     @Override
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
index 51d0c6cf..47fe9710 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
@@ -26,6 +26,8 @@ import net.md_5.bungee.protocol.PacketWrapper;
 import net.md_5.bungee.protocol.ProtocolConstants;
 import net.md_5.bungee.protocol.packet.Chat;
 import net.md_5.bungee.protocol.packet.ClientSettings;
+import net.md_5.bungee.protocol.packet.EntityEffect;
+import net.md_5.bungee.protocol.packet.EntityRemoveEffect;
 import net.md_5.bungee.protocol.packet.KeepAlive;
 import net.md_5.bungee.protocol.packet.PlayerListItem;
 import net.md_5.bungee.protocol.packet.PluginMessage;
@@ -266,6 +268,22 @@ public class UpstreamBridge extends PacketHandler
         }
     }
 
+    // IvanCord start - handle upstream EntityEffect and EntityRemoveEffect packets
+    // They were not handled by waterfall
+    @Override
+    public void handle(EntityEffect entityEffect)
+    {
+        con.getPotions().add( entityEffect.getEffectId() );
+    }
+
+    @Override
+    public void handle(EntityRemoveEffect entityRemoveEffect)
+    {
+        con.getPotions().remove( entityRemoveEffect.getEffectId() );
+    }
+
+    // IvanCord end
+
     @Override
     public String toString()
     {
diff --git a/proxy/src/main/java/net/md_5/bungee/forge/ForgeClientHandler.java b/proxy/src/main/java/net/md_5/bungee/forge/ForgeClientHandler.java
index c1272da3..eda5566d 100644
--- a/proxy/src/main/java/net/md_5/bungee/forge/ForgeClientHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/forge/ForgeClientHandler.java
@@ -105,9 +105,12 @@ public class ForgeClientHandler
 
     private void resetAllThePotions(UserConnection con) {
         // Just to be sure
-        for (Map.Entry<Integer, Integer> entry: con.getPotions().entries()) {
-            con.unsafe().sendPacket(new EntityRemoveEffect(entry.getKey(), entry.getValue()));
+        // IvanCord start - rewrite this
+        for ( int potion : con.getPotions() )
+        {
+            con.unsafe().sendPacket( new EntityRemoveEffect( con.getClientEntityId(), potion ) );
         }
+        // IvanCord end
         con.getPotions().clear();
     }
 
-- 
2.21.0.windows.1

