From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Sun, 14 Apr 2019 08:54:05 +0300
Subject: [PATCH] Add IvanCord specific configuration


diff --git a/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java b/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
new file mode 100644
index 0000000000000000000000000000000000000000..06d74844fe24fb7bdc33dc81c8ec28280e7a1c12
--- /dev/null
+++ b/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
@@ -0,0 +1,18 @@
+package com.mrivanplays.ivancord.conf;
+
+import java.io.File;
+
+import io.github.waterfallmc.waterfall.conf.WaterfallConfiguration;
+import net.md_5.bungee.conf.YamlConfig;
+
+public class IvanCordConfiguration extends WaterfallConfiguration
+{
+
+    @Override
+    public void load()
+    {
+        super.load();
+        YamlConfig config = new YamlConfig( new File( "ivancord.yml" ) );
+        config.load( false );
+    }
+}
diff --git a/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java b/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
index e31fb00ae9de9864c8c3c52ffa9eb9ee4c0692a1..99a69f7d24464ee92aa8008ed3939f9f8897df26 100644
--- a/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
+++ b/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
@@ -7,7 +7,7 @@ import net.md_5.bungee.protocol.ProtocolConstants;
 
 import java.io.File;
 
-public class WaterfallConfiguration extends Configuration {
+public abstract class WaterfallConfiguration extends Configuration { // IvanCord - abstract
 
     /**
      * Whether we log InitialHandler connections
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 2988448ba1534c9f819bd7a4c7b644dce8cc7e4d..083c7f81cae68c0c578dcb7797b8e67a966b8733 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -117,7 +117,7 @@ public class BungeeCord extends ProxyServer
      * Configuration.
      */
     @Getter
-    public final Configuration config = new WaterfallConfiguration();
+    public final Configuration config = new com.mrivanplays.ivancord.conf.IvanCordConfiguration(); // IvanCord - ivancord specific configuration
     /**
      * Localization bundle.
      */
