From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Sun, 14 Apr 2019 08:54:05 +0300
Subject: [PATCH] Add IvanCord specific configuration


diff --git a/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java b/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
new file mode 100644
index 0000000000000000000000000000000000000000..06d74844fe24fb7bdc33dc81c8ec28280e7a1c12
--- /dev/null
+++ b/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
@@ -0,0 +1,18 @@
+package com.mrivanplays.ivancord.conf;
+
+import java.io.File;
+
+import io.github.waterfallmc.waterfall.conf.WaterfallConfiguration;
+import net.md_5.bungee.conf.YamlConfig;
+
+public class IvanCordConfiguration extends WaterfallConfiguration
+{
+
+    @Override
+    public void load()
+    {
+        super.load();
+        YamlConfig config = new YamlConfig( new File( "ivancord.yml" ) );
+        config.load( false );
+    }
+}
diff --git a/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java b/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
index da0efa36f4fdb5bcfa652d67b9526a3d88dc92fa..8977aaf3836f18cea4f2f585363470eea30bd761 100644
--- a/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
+++ b/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
@@ -7,7 +7,7 @@ import net.md_5.bungee.protocol.ProtocolConstants;
 
 import java.io.File;
 
-public class WaterfallConfiguration extends Configuration {
+public abstract class WaterfallConfiguration extends Configuration { // IvanCord - abstract
 
     /**
      * Whether we log InitialHandler connections
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 8c08197555fee3eccc01c19bb0e11189952c7a3a..75c6ff5be7b166501adb6dd88df3df568c3adc19 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -110,7 +110,7 @@ public class BungeeCord extends ProxyServer
      * Configuration.
      */
     @Getter
-    public final Configuration config = new WaterfallConfiguration();
+    public final Configuration config = new com.mrivanplays.ivancord.conf.IvanCordConfiguration(); // IvanCord - ivancord specific configuration
     /**
      * Localization formats.
      */
