From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Mon, 22 Apr 2019 10:45:15 +0300
Subject: [PATCH] Kick to lobby support


diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
index d69463f07cc15a58bc5b6a6cf016595db95723ca..6f91deb32a5a9191340638431c1480dece5392a8 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
@@ -261,4 +261,15 @@ public interface ProxyConfig
      * @return {@code true} if tablist rewriting is disabled, {@code false} otherwise
      */
     boolean isDisableTabListRewrite();
+
+    // IvanCord start
+
+    /**
+     * If this is true, a listener to {@link net.md_5.bungee.api.event.ServerKickEvent} will be
+     * registered, which will send a kicked player from x server to the first server in {@link ListenerInfo#getServerPriority()}
+     *
+     * @return should the player kicked be transferred to lobby
+     */
+    boolean isKickToLobby();
+    // IvanCord end
 }
diff --git a/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java b/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
index 06d74844fe24fb7bdc33dc81c8ec28280e7a1c12..238bc3c1f43aa5bd6de34f4355ede2dee818fef3 100644
--- a/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
+++ b/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
@@ -3,16 +3,21 @@ package com.mrivanplays.ivancord.conf;
 import java.io.File;
 
 import io.github.waterfallmc.waterfall.conf.WaterfallConfiguration;
+import lombok.Getter;
 import net.md_5.bungee.conf.YamlConfig;
 
+@Getter
 public class IvanCordConfiguration extends WaterfallConfiguration
 {
 
+    private boolean kickToLobby = true;
+
     @Override
     public void load()
     {
         super.load();
         YamlConfig config = new YamlConfig( new File( "ivancord.yml" ) );
         config.load( false );
+        kickToLobby = config.getBoolean( "kick_to_lobby", kickToLobby );
     }
 }
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index 9e180c30c05d946e892ae9fbd4d7d796e202d5d4..2c2427a6c2acf037cbb84b3a440e92cb64519352 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -31,6 +31,7 @@ import net.md_5.bungee.ServerConnection.KeepAliveData;
 import net.md_5.bungee.UserConnection;
 import net.md_5.bungee.Util;
 import net.md_5.bungee.api.ProxyServer;
+import net.md_5.bungee.api.chat.ComponentBuilder; // IvanCord
 import net.md_5.bungee.api.chat.BaseComponent;
 import net.md_5.bungee.api.chat.TextComponent;
 import net.md_5.bungee.api.config.ServerInfo;
@@ -569,14 +570,46 @@ public class DownstreamBridge extends PacketHandler
         {
             def = null;
         }
-        ServerKickEvent event = bungee.getPluginManager().callEvent( new ServerKickEvent( con, server.getInfo(), ComponentSerializer.parse( kick.getMessage() ), def, ServerKickEvent.State.CONNECTED, ServerKickEvent.Cause.SERVER ) ); // Waterfall
-        if ( event.isCancelled() && event.getCancelServer() != null )
+        // IvanCord start - kick to lobby
+        boolean kickToLobby = bungee.getConfig().isKickToLobby();
+        ServerInfo defaultServer = bungee.getServerInfo( bungee.getConfig().getListeners().parallelStream().findFirst().get().getDefaultServer() );
+        ServerKickEvent event = bungee.getPluginManager().callEvent( new ServerKickEvent( con, server.getInfo(),
+                ComponentSerializer.parse( kick.getMessage() ), def, ServerKickEvent.State.CONNECTED, ServerKickEvent.Cause.SERVER ) );
+        BaseComponent[] kickMessage = new ComponentBuilder( bungee.getTranslation( "server_kick" ) )
+                .append( event.getKickReasonComponent() )
+                .create();
+        if ( kickToLobby )
         {
-            con.connectNow( event.getCancelServer(), ServerConnectEvent.Reason.KICK_REDIRECT );
+            if ( defaultServer != null )
+            {
+                con.connectNow( defaultServer, ServerConnectEvent.Reason.KICK_REDIRECT );
+                con.sendMessage( TextComponent.fromLegacyText(
+                        bungee.getTranslation( "kick_to_lobby", event.getKickedFrom().getName(),
+                                BaseComponent.toLegacyText( kickMessage ), defaultServer.getName() ) ) );
+            } else
+            {
+                if ( event.isCancelled() && event.getCancelServer() != null )
+                {
+                    con.connectNow( event.getCancelServer(), ServerConnectEvent.Reason.KICK_REDIRECT );
+                    con.sendMessage( TextComponent.fromLegacyText(
+                            bungee.getTranslation( "kick_to_lobby", event.getKickedFrom().getName(),
+                                    BaseComponent.toLegacyText( kickMessage ), event.getCancelServer().getName() ) ) );
+                } else
+                {
+                    con.disconnect0( kickMessage );
+                }
+            }
         } else
         {
-            con.disconnect0( event.getKickReasonComponent() ); // TODO: Prefix our own stuff.
+            if ( event.isCancelled() && event.getCancelServer() != null )
+            {
+                con.connectNow( event.getCancelServer(), ServerConnectEvent.Reason.KICK_REDIRECT );
+            } else
+            {
+                con.disconnect0( kickMessage );
+            }
         }
+        // IvanCord end
         server.setObsolete( true );
         throw CancelSendSignal.INSTANCE;
     }
diff --git a/proxy/src/main/resources/messages.properties b/proxy/src/main/resources/messages.properties
index 2a79e62b65a6e1f964793f732836a01bb6ee4ec4..e838024539689e7ff1109d116434ddc64c9e088f 100644
--- a/proxy/src/main/resources/messages.properties
+++ b/proxy/src/main/resources/messages.properties
@@ -38,3 +38,4 @@ you_got_summoned=\u00a76Summoned to {0} by {1}
 command_perms_groups=\u00a76You have the following groups: {0}
 command_perms_permission=\u00a79- {0}
 command_ip=\u00a79IP of {0} is {1}
+kick_to_lobby=You were kicked from server {0} with reason {1} . The fallback sent you to {2}
