From c09fd54af2183e8575fc10fcfb5b6367c952d674 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Mon, 27 May 2019 14:25:44 +0300
Subject: [PATCH] Speed up packet construction


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
index 652e4ca2b..e830aa158 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
@@ -7,6 +7,7 @@ import gnu.trove.map.TObjectIntMap;
 import gnu.trove.map.hash.TIntObjectHashMap;
 import gnu.trove.map.hash.TObjectIntHashMap;
 import java.lang.reflect.Constructor;
+import java.util.function.Supplier;
 import lombok.Data;
 import lombok.Getter;
 import net.md_5.bungee.protocol.packet.BossBar;
@@ -58,6 +59,7 @@ public enum Protocol
         {
             TO_SERVER.registerPacket(
                     Handshake.class,
+                    Handshake::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x00 )
             );
         }
@@ -69,6 +71,7 @@ public enum Protocol
         {
             TO_CLIENT.registerPacket(
                     KeepAlive.class,
+                    KeepAlive::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x00 ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x1F ),
                     map( ProtocolConstants.MINECRAFT_1_13, 0x21 ),
@@ -76,18 +79,21 @@ public enum Protocol
             );
             TO_CLIENT.registerPacket(
                     Login.class,
+                    Login::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x01 ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x23 ),
                     map( ProtocolConstants.MINECRAFT_1_13, 0x25 )
             );
             TO_CLIENT.registerPacket(
                     Chat.class,
+                    Chat::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x02 ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x0F ),
                     map( ProtocolConstants.MINECRAFT_1_13, 0x0E )
             );
             TO_CLIENT.registerPacket(
                     Respawn.class,
+                    Respawn::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x07 ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x33 ),
                     map( ProtocolConstants.MINECRAFT_1_12, 0x34 ),
@@ -97,11 +103,13 @@ public enum Protocol
             );
             TO_CLIENT.registerPacket(
                     BossBar.class,
+                    BossBar::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_9, 0x0C )
             );
             // Waterfall start
             TO_CLIENT.registerPacket(
                     EntityEffect.class,
+                    EntityEffect::new, // IvanCord - speed up packet construction
                     map(ProtocolConstants.MINECRAFT_1_8, 0x1D),
                     map(ProtocolConstants.MINECRAFT_1_9, 0x4C),
                     map(ProtocolConstants.MINECRAFT_1_9_4, 0x4B),
@@ -112,6 +120,7 @@ public enum Protocol
             );
             TO_CLIENT.registerPacket(
                     EntityRemoveEffect.class,
+                    EntityRemoveEffect::new, // IvanCord - speed up packet construction
                     map(ProtocolConstants.MINECRAFT_1_8, 0x1E),
                     map(ProtocolConstants.MINECRAFT_1_9, 0x31),
                     map(ProtocolConstants.MINECRAFT_1_12, 0x32),
@@ -122,6 +131,7 @@ public enum Protocol
             // Waterfall end
             TO_CLIENT.registerPacket(
                     PlayerListItem.class, // PlayerInfo
+                    PlayerListItem::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x38 ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x2D ),
                     map( ProtocolConstants.MINECRAFT_1_12_1, 0x2E ),
@@ -130,6 +140,7 @@ public enum Protocol
             );
             TO_CLIENT.registerPacket(
                     TabCompleteResponse.class,
+                    TabCompleteResponse::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x3A ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x0E ),
                     map( ProtocolConstants.MINECRAFT_1_13, 0x10 )
@@ -138,6 +149,7 @@ public enum Protocol
             /*
             TO_CLIENT.registerPacket(
                     ScoreboardObjective.class,
+                    ScoreboardObjective::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x3B ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x3F ),
                     map( ProtocolConstants.MINECRAFT_1_12, 0x41 ),
@@ -147,6 +159,7 @@ public enum Protocol
             );
             TO_CLIENT.registerPacket(
                     ScoreboardScore.class,
+                    ScoreboardScore::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x3C ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x42 ),
                     map( ProtocolConstants.MINECRAFT_1_12, 0x44 ),
@@ -156,6 +169,7 @@ public enum Protocol
             );
             TO_CLIENT.registerPacket(
                     ScoreboardDisplay.class,
+                    ScoreboardDisplay::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x3D ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x38 ),
                     map( ProtocolConstants.MINECRAFT_1_12, 0x3A ),
@@ -165,6 +179,7 @@ public enum Protocol
             );
             TO_CLIENT.registerPacket(
                     Team.class,
+                    Team::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x3E ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x41 ),
                     map( ProtocolConstants.MINECRAFT_1_12, 0x43 ),
@@ -176,6 +191,7 @@ public enum Protocol
             // IvanCord end
             TO_CLIENT.registerPacket(
                     PluginMessage.class,
+                    PluginMessage::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x3F ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x18 ),
                     map( ProtocolConstants.MINECRAFT_1_13, 0x19 ),
@@ -183,6 +199,7 @@ public enum Protocol
             );
             TO_CLIENT.registerPacket(
                     Kick.class,
+                    Kick::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x40 ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x1A ),
                     map( ProtocolConstants.MINECRAFT_1_13, 0x1B ),
@@ -190,6 +207,7 @@ public enum Protocol
             );
             TO_CLIENT.registerPacket(
                     Title.class,
+                    Title::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x45 ),
                     map( ProtocolConstants.MINECRAFT_1_12, 0x47 ),
                     map( ProtocolConstants.MINECRAFT_1_12_1, 0x48 ),
@@ -198,6 +216,7 @@ public enum Protocol
             );
             TO_CLIENT.registerPacket(
                     PlayerListHeaderFooter.class,
+                    PlayerListHeaderFooter::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x47 ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x48 ),
                     map( ProtocolConstants.MINECRAFT_1_9_4, 0x47 ),
@@ -208,6 +227,7 @@ public enum Protocol
             );
             TO_CLIENT.registerPacket(
                     EntityStatus.class,
+                    EntityStatus::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x1A ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x1B ),
                     map( ProtocolConstants.MINECRAFT_1_13, 0x1C ),
@@ -215,15 +235,18 @@ public enum Protocol
             );
             TO_CLIENT.registerPacket(
                     Commands.class,
+                    Commands::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_13, 0x11 )
             );
             TO_CLIENT.registerPacket(
                     ViewDistance.class,
+                    ViewDistance::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_14, 0x41 )
             );
 
             TO_SERVER.registerPacket(
                     KeepAlive.class,
+                    KeepAlive::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x00 ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x0B ),
                     map( ProtocolConstants.MINECRAFT_1_12, 0x0C ),
@@ -233,6 +256,7 @@ public enum Protocol
             );
             TO_SERVER.registerPacket(
                     Chat.class,
+                    Chat::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x01 ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x02 ),
                     map( ProtocolConstants.MINECRAFT_1_12, 0x03 ),
@@ -241,6 +265,7 @@ public enum Protocol
             );
             TO_SERVER.registerPacket(
                     TabCompleteRequest.class,
+                    TabCompleteRequest::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x14 ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x01 ),
                     map( ProtocolConstants.MINECRAFT_1_12, 0x02 ),
@@ -250,6 +275,7 @@ public enum Protocol
             );
             TO_SERVER.registerPacket(
                     ClientSettings.class,
+                    ClientSettings::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x15 ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x04 ),
                     map( ProtocolConstants.MINECRAFT_1_12, 0x05 ),
@@ -258,6 +284,7 @@ public enum Protocol
             );
             TO_SERVER.registerPacket(
                     PluginMessage.class,
+                    PluginMessage::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x17 ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x09 ),
                     map( ProtocolConstants.MINECRAFT_1_12, 0x0A ),
@@ -274,19 +301,23 @@ public enum Protocol
         {
             TO_CLIENT.registerPacket(
                     StatusResponse.class,
+                    StatusResponse::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x00 )
             );
             TO_CLIENT.registerPacket(
                     PingPacket.class,
+                    PingPacket::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x01 )
             );
 
             TO_SERVER.registerPacket(
                     StatusRequest.class,
+                    StatusRequest::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x00 )
             );
             TO_SERVER.registerPacket(
                     PingPacket.class,
+                    PingPacket::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x01 )
             );
         }
@@ -298,35 +329,43 @@ public enum Protocol
         {
             TO_CLIENT.registerPacket(
                     Kick.class,
+                    Kick::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x00 )
             );
             TO_CLIENT.registerPacket(
                     EncryptionRequest.class,
+                    EncryptionRequest::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x01 )
             );
             TO_CLIENT.registerPacket(
                     LoginSuccess.class,
+                    LoginSuccess::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x02 )
             );
             TO_CLIENT.registerPacket(
                     SetCompression.class,
+                    SetCompression::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x03 )
             );
             TO_CLIENT.registerPacket(
                     LoginPayloadRequest.class,
+                    LoginPayloadRequest::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_13, 0x04 )
             );
 
             TO_SERVER.registerPacket(
                     LoginRequest.class,
+                    LoginRequest::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x00 )
             );
             TO_SERVER.registerPacket(
                     EncryptionResponse.class,
+                    EncryptionResponse::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_8, 0x01 )
             );
             TO_SERVER.registerPacket(
                     LoginPayloadResponse.class,
+                    LoginPayloadResponse::new, // IvanCord - speed up packet construction
                     map( ProtocolConstants.MINECRAFT_1_13, 0x02 )
             );
         }
@@ -377,7 +416,7 @@ public enum Protocol
 
         private final int protocolVersion;
         private final TObjectIntMap<Class<? extends DefinedPacket>> packetMap = new TObjectIntHashMap<>( MAX_PACKET_ID );
-        private final Constructor<? extends DefinedPacket>[] packetConstructors = new Constructor[ MAX_PACKET_ID ];
+        private final Supplier<? extends DefinedPacket>[] packetConstructors = new Supplier[ MAX_PACKET_ID ]; // IvanCord - speed up packet construction
     }
 
     @Data
@@ -445,21 +484,25 @@ public enum Protocol
                 throw new BadPacketException( "Packet with id " + id + " outside of range " );
             }
 
-            Constructor<? extends DefinedPacket> constructor = protocolData.packetConstructors[id];
+            // IvanCord start - speed up packet construction
+            Supplier<? extends DefinedPacket> constructor = protocolData.packetConstructors[id];
             try
             {
-                return ( constructor == null ) ? null : constructor.newInstance();
-            } catch ( ReflectiveOperationException ex )
+                return ( constructor == null ) ? null : constructor.get();
+            } catch ( Throwable ex )
             {
+                // IvanCord end
                 throw new BadPacketException( "Could not construct packet with id " + id, ex );
             }
         }
 
-        private void registerPacket(Class<? extends DefinedPacket> packetClass, ProtocolMapping... mappings)
+        // IvanCord start - speed up packet construction
+        private <T extends DefinedPacket> void registerPacket(Class<T> packetClass, Supplier<T> constructor, ProtocolMapping... mappings)
         {
             try
             {
-                Constructor<? extends DefinedPacket> constructor = packetClass.getDeclaredConstructor();
+                // Constructor<? extends DefinedPacket> constructor = packetClass.getDeclaredConstructor();
+                // IvanCord end
 
                 int mappingIndex = 0;
                 ProtocolMapping mapping = mappings[mappingIndex];
@@ -488,11 +531,36 @@ public enum Protocol
                     data.packetMap.put( packetClass, mapping.packetID );
                     data.packetConstructors[mapping.packetID] = constructor;
                 }
-            } catch ( NoSuchMethodException ex )
+            } catch ( Throwable ex ) // IvanCord - speed up packet construction
+            {
+                throw new BadPacketException( "No NoArgsConstructor for packet class " + packetClass );
+            }
+        }
+
+        // IvanCord start - add old construction
+        private <T extends DefinedPacket> void registerPacket(Class<T> packetClass, ProtocolMapping... mappings)
+        {
+            Supplier<T> packetSupplier;
+            try
+            {
+                Constructor<? extends DefinedPacket> constructor = packetClass.getDeclaredConstructor();
+                packetSupplier = () ->
+                {
+                    try
+                    {
+                        return (T) constructor.newInstance();
+                    } catch ( Throwable e )
+                    {
+                        throw new RuntimeException( e );
+                    }
+                };
+            } catch ( ReflectiveOperationException e )
             {
                 throw new BadPacketException( "No NoArgsConstructor for packet class " + packetClass );
             }
+            registerPacket( packetClass, packetSupplier, mappings );
         }
+        // IvanCord end
 
         final int getId(Class<? extends DefinedPacket> packet, int version)
         {
-- 
2.22.0.windows.1

