From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Wed, 12 Jun 2019 20:04:42 +0300
Subject: [PATCH] Add player positioning api


diff --git a/api/src/main/java/com/mrivanplays/ivancord/api/Position.java b/api/src/main/java/com/mrivanplays/ivancord/api/Position.java
new file mode 100644
index 0000000000000000000000000000000000000000..cc45311629bcb92bb8d93ec43e422740487e08cd
--- /dev/null
+++ b/api/src/main/java/com/mrivanplays/ivancord/api/Position.java
@@ -0,0 +1,47 @@
+package com.mrivanplays.ivancord.api;
+
+import lombok.AllArgsConstructor;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+
+/**
+ * Represents a player position
+ */
+@Getter
+@AllArgsConstructor
+@ToString
+@EqualsAndHashCode
+public class Position
+{
+
+    /**
+     * X axis
+     */
+    private double x;
+
+    /**
+     * Y axis
+     */
+    private double y;
+
+    /**
+     * Z axis
+     */
+    private double z;
+
+    /**
+     * Yaw axis
+     */
+    private float yaw;
+
+    /**
+     * Pitch axis
+     */
+    private float pitch;
+
+    /**
+     * Does the player is on ground
+     */
+    private boolean onGround;
+}
diff --git a/api/src/main/java/com/mrivanplays/ivancord/api/event/PlayerChangePositionEvent.java b/api/src/main/java/com/mrivanplays/ivancord/api/event/PlayerChangePositionEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..cf92ce1221e8d7b19f7b4f12e6e1d6393eb897e2
--- /dev/null
+++ b/api/src/main/java/com/mrivanplays/ivancord/api/event/PlayerChangePositionEvent.java
@@ -0,0 +1,46 @@
+package com.mrivanplays.ivancord.api.event;
+
+import com.mrivanplays.ivancord.api.Position;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+import net.md_5.bungee.api.connection.ProxiedPlayer;
+import net.md_5.bungee.api.plugin.Cancellable;
+import net.md_5.bungee.api.plugin.Event;
+
+/**
+ * Event, called when player is about to change his position.
+ */
+@ToString
+@EqualsAndHashCode(callSuper = false)
+@Getter
+public class PlayerChangePositionEvent extends Event implements Cancellable
+{
+
+    @Setter
+    private boolean cancelled;
+
+    /**
+     * The {@link ProxiedPlayer}, which changed position.
+     */
+    private ProxiedPlayer player;
+
+    /**
+     * The position the {@link ProxiedPlayer} moved from.
+     */
+    private Position movedFrom;
+
+    /**
+     * The position the {@link ProxiedPlayer} moved to.
+     */
+    @Setter
+    private Position movedTo;
+
+    public PlayerChangePositionEvent(ProxiedPlayer player, Position movedFrom, Position movedTo)
+    {
+        this.player = player;
+        this.movedFrom = movedFrom;
+        this.movedTo = movedTo;
+    }
+}
diff --git a/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java b/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java
index b14463bd6864054f38988d7bdaaa105635a3a263..03bc8c4b0cfa9733b93e74546a51bca1b15bca44 100644
--- a/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java
+++ b/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java
@@ -13,6 +13,7 @@ import net.md_5.bungee.api.chat.BaseComponent;
 import net.md_5.bungee.api.config.ServerInfo;
 import net.md_5.bungee.api.event.ServerConnectEvent;
 import net.md_5.bungee.api.score.Scoreboard;
+import com.mrivanplays.ivancord.api.Position; // IvanCord
 
 /**
  * Represents a player who's connection is being connected to somewhere else,
@@ -283,6 +284,16 @@ public interface ProxiedPlayer extends Connection, CommandSender
      */
     byte getViewDistance();
 
+    // IvanCord start - add player position api
+
+    /**
+     * Gets the player's position
+     *
+     * @return player position
+     */
+    Position getPosition();
+    // IvanCord end
+
     /**
      * Gets this player's chat mode.
      *
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/AbstractPacketHandler.java b/protocol/src/main/java/net/md_5/bungee/protocol/AbstractPacketHandler.java
index ee2317e8402ee98da9456a9b43e4f6d53753cb64..89b3fdbd320871d41b2dfdc670e96bb5cae4a0a3 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/AbstractPacketHandler.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/AbstractPacketHandler.java
@@ -25,6 +25,7 @@ import net.md_5.bungee.protocol.packet.LoginSuccess;
 import net.md_5.bungee.protocol.packet.PingPacket;
 import net.md_5.bungee.protocol.packet.PlayerListHeaderFooter;
 import net.md_5.bungee.protocol.packet.PlayerListItem;
+import net.md_5.bungee.protocol.packet.PlayerPosition; // IvanCord
 import net.md_5.bungee.protocol.packet.PluginMessage;
 import net.md_5.bungee.protocol.packet.Respawn;
 import net.md_5.bungee.protocol.packet.ScoreboardDisplay;
@@ -196,6 +197,12 @@ public abstract class AbstractPacketHandler
     {
     }
 
+    // IvanCord start
+    public void handle(PlayerPosition playerPosition) throws Exception
+    {
+    }
+    // IvanCord end
+
     // Waterfall start
     public void handle(EntityEffect entityEffect) throws Exception
     {
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
index 416559c85140eb93c311f5b2c51c12699301cdba..6d3d5d23c3eb7933f7c886776db1e3379c4768e1 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
@@ -31,6 +31,7 @@ import net.md_5.bungee.protocol.packet.LoginSuccess;
 import net.md_5.bungee.protocol.packet.PingPacket;
 import net.md_5.bungee.protocol.packet.PlayerListHeaderFooter;
 import net.md_5.bungee.protocol.packet.PlayerListItem;
+import net.md_5.bungee.protocol.packet.PlayerPosition;
 import net.md_5.bungee.protocol.packet.PluginMessage;
 import net.md_5.bungee.protocol.packet.Respawn;
 import net.md_5.bungee.protocol.packet.ScoreboardDisplay;
@@ -369,6 +370,19 @@ public enum Protocol
                     map( ProtocolConstants.MINECRAFT_1_14, 0x0B ),
                     map( ProtocolConstants.MINECRAFT_1_17, 0x0A )
             );
+            // IvanCord start - add player position
+            TO_SERVER.registerPacket(
+                    PlayerPosition.class,
+                    PlayerPosition::new,
+                    map( ProtocolConstants.MINECRAFT_1_8, 0x06 ),
+                    map( ProtocolConstants.MINECRAFT_1_9, 0x0D ),
+                    map( ProtocolConstants.MINECRAFT_1_12, 0x0F ),
+                    map( ProtocolConstants.MINECRAFT_1_12_1, 0x0E ),
+                    map( ProtocolConstants.MINECRAFT_1_13, 0x11 ),
+                    map( ProtocolConstants.MINECRAFT_1_14, 0x12 ),
+                    map( ProtocolConstants.MINECRAFT_1_16, 0x13 )
+            );
+            // IvanCord end
         }
     },
     // 1
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/packet/PlayerPosition.java b/protocol/src/main/java/net/md_5/bungee/protocol/packet/PlayerPosition.java
new file mode 100644
index 0000000000000000000000000000000000000000..f23fe50e8c9a22ce934f5797f9e75fb2767a9479
--- /dev/null
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/packet/PlayerPosition.java
@@ -0,0 +1,54 @@
+// IvanCord start
+package net.md_5.bungee.protocol.packet;
+
+import io.netty.buffer.ByteBuf;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.NoArgsConstructor;
+import net.md_5.bungee.protocol.AbstractPacketHandler;
+import net.md_5.bungee.protocol.DefinedPacket;
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Data
+@EqualsAndHashCode(callSuper = false)
+public class PlayerPosition extends DefinedPacket
+{
+
+    private double x;
+    private double y;
+    private double z;
+    private float yaw;
+    private float pitch;
+    private boolean onGround;
+
+    @Override
+    public void read(ByteBuf buf)
+    {
+        x = buf.readDouble();
+        y = buf.readDouble();
+        z = buf.readDouble();
+        yaw = buf.readFloat();
+        pitch = buf.readFloat();
+        onGround = buf.readBoolean();
+    }
+
+    @Override
+    public void write(ByteBuf buf)
+    {
+        buf.writeDouble( x );
+        buf.writeDouble( y );
+        buf.writeDouble( z );
+        buf.writeFloat( yaw );
+        buf.writeFloat( pitch );
+        buf.writeBoolean( onGround );
+    }
+
+    @Override
+    public void handle(AbstractPacketHandler handler) throws Exception
+    {
+        handler.handle( this );
+    }
+}
+// IvanCord end
diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 4641de8796eebcaa81226c2a88cc08d2e35dbffb..a2a37ab133d9781e69bee98ac8ce03d70c541899 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -66,6 +66,7 @@ import net.md_5.bungee.tab.ServerUnique;
 import net.md_5.bungee.tab.TabList;
 import net.md_5.bungee.util.CaseInsensitiveSet;
 import net.md_5.bungee.util.ChatComponentTransformer;
+import com.mrivanplays.ivancord.api.Position; // IvanCord
 
 @RequiredArgsConstructor
 public final class UserConnection implements ProxiedPlayer
@@ -106,6 +107,10 @@ public final class UserConnection implements ProxiedPlayer
     @Getter
     @Setter
     private int gamemode;
+    // IvanCord start
+    @Getter
+    private Position position;
+    // IvanCord end
     @Getter
     private int compressionThreshold = -1;
     // Used for trying multiple servers in order
@@ -661,6 +666,12 @@ public final class UserConnection implements ProxiedPlayer
     {
         return ( settings != null ) ? settings.getViewDistance() : 10;
     }
+    // IvanCord start
+    public void setPosition(Position position)
+    {
+        this.position = position;
+    }
+    // IvanCord end
 
     @Override
     public ProxiedPlayer.ChatMode getChatMode()
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index 0b7a7b4d880449a95fe2a5cf99844b5e7bc67144..409878ec8c403062d61a50ef3cb0f03bd138bb13 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -43,7 +43,7 @@ import net.md_5.bungee.api.event.ServerKickEvent;
 import net.md_5.bungee.api.event.TabCompleteResponseEvent;
 import net.md_5.bungee.api.plugin.Command;
 import net.md_5.bungee.api.score.Objective;
-import net.md_5.bungee.api.score.Position;
+//import net.md_5.bungee.api.score.Position; // IvanCord - comment out for player positioning api
 import net.md_5.bungee.api.score.Score;
 import net.md_5.bungee.api.score.Scoreboard;
 import net.md_5.bungee.api.score.Team;
@@ -61,6 +61,7 @@ import net.md_5.bungee.protocol.packet.EntityRemoveEffect;
 import net.md_5.bungee.protocol.packet.KeepAlive;
 import net.md_5.bungee.protocol.packet.Kick;
 import net.md_5.bungee.protocol.packet.PlayerListItem;
+import net.md_5.bungee.protocol.packet.PlayerPosition; // IvanCord
 import net.md_5.bungee.protocol.packet.PluginMessage;
 import net.md_5.bungee.protocol.packet.Respawn;
 import net.md_5.bungee.protocol.packet.ScoreboardDisplay;
@@ -69,6 +70,10 @@ import net.md_5.bungee.protocol.packet.ScoreboardScore;
 import net.md_5.bungee.protocol.packet.SetCompression;
 import net.md_5.bungee.protocol.packet.TabCompleteResponse;
 import net.md_5.bungee.tab.TabList;
+// IvanCord start
+import com.mrivanplays.ivancord.api.Position;
+import com.mrivanplays.ivancord.api.event.PlayerChangePositionEvent;
+// IvanCord end
 
 @RequiredArgsConstructor
 public class DownstreamBridge extends PacketHandler
@@ -768,6 +773,39 @@ public class DownstreamBridge extends PacketHandler
         }
     }
 
+    // IvanCord start
+    @Override
+    public void handle(PlayerPosition pac)
+    {
+        Position packetPos = new Position( pac.getX(), pac.getY(), pac.getZ(), pac.getYaw(), pac.getPitch(), pac.isOnGround() );
+        Position conPos = con.getPosition();
+        boolean changed = true;
+        if ( packetPos.equals( conPos ) )
+        {
+            changed = false;
+        }
+        if ( changed )
+        {
+            PlayerChangePositionEvent event = new PlayerChangePositionEvent( con, conPos, packetPos );
+            if ( event.callEvent() )
+            {
+                con.unsafe().sendPacket( pac );
+            } else
+            {
+                if ( !event.getMovedTo().equals( packetPos ) )
+                {
+                    Position pn = event.getMovedTo();
+                    con.unsafe().sendPacket( new PlayerPosition( pn.getX(), pn.getY(), pn.getZ(), pn.getYaw(), pn.getPitch(), pn.isOnGround() ) );
+                } else
+                {
+                    con.unsafe().sendPacket( pac );
+                }
+            }
+            throw CancelSendSignal.INSTANCE;
+        }
+    }
+    // IvanCord end
+
     @Override
     public String toString()
     {
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
index 60512bd2f74e465f68dddd9bdc81a4a4302be633..30e6cc6552a134bc5358414cbc93368f9bf63ca0 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
@@ -29,9 +29,11 @@ import net.md_5.bungee.protocol.packet.Chat;
 import net.md_5.bungee.protocol.packet.ClientSettings;
 import net.md_5.bungee.protocol.packet.KeepAlive;
 import net.md_5.bungee.protocol.packet.PlayerListItem;
+import net.md_5.bungee.protocol.packet.PlayerPosition;
 import net.md_5.bungee.protocol.packet.PluginMessage;
 import net.md_5.bungee.protocol.packet.TabCompleteRequest;
 import net.md_5.bungee.protocol.packet.TabCompleteResponse;
+import com.mrivanplays.ivancord.api.Position; // IvanCord
 
 public class UpstreamBridge extends PacketHandler
 {
@@ -287,6 +289,14 @@ public class UpstreamBridge extends PacketHandler
         }
     }
 
+    // IvanCord start
+    @Override
+    public void handle(PlayerPosition pac)
+    {
+        con.setPosition( new Position( pac.getX(), pac.getY(), pac.getZ(), pac.getYaw(), pac.getPitch(), pac.isOnGround() ) );
+    }
+    // IvanCord end
+
     @Override
     public String toString()
     {
