From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Wed, 12 Jun 2019 20:04:42 +0300
Subject: [PATCH] Add player positioning api


diff --git a/api/src/main/java/com/mrivanplays/ivancord/api/Position.java b/api/src/main/java/com/mrivanplays/ivancord/api/Position.java
new file mode 100644
index 0000000000000000000000000000000000000000..cc45311629bcb92bb8d93ec43e422740487e08cd
--- /dev/null
+++ b/api/src/main/java/com/mrivanplays/ivancord/api/Position.java
@@ -0,0 +1,47 @@
+package com.mrivanplays.ivancord.api;
+
+import lombok.AllArgsConstructor;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+
+/**
+ * Represents a player position
+ */
+@Getter
+@AllArgsConstructor
+@ToString
+@EqualsAndHashCode
+public class Position
+{
+
+    /**
+     * X axis
+     */
+    private double x;
+
+    /**
+     * Y axis
+     */
+    private double y;
+
+    /**
+     * Z axis
+     */
+    private double z;
+
+    /**
+     * Yaw axis
+     */
+    private float yaw;
+
+    /**
+     * Pitch axis
+     */
+    private float pitch;
+
+    /**
+     * Does the player is on ground
+     */
+    private boolean onGround;
+}
diff --git a/api/src/main/java/com/mrivanplays/ivancord/api/event/PlayerChangePositionEvent.java b/api/src/main/java/com/mrivanplays/ivancord/api/event/PlayerChangePositionEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..cf92ce1221e8d7b19f7b4f12e6e1d6393eb897e2
--- /dev/null
+++ b/api/src/main/java/com/mrivanplays/ivancord/api/event/PlayerChangePositionEvent.java
@@ -0,0 +1,46 @@
+package com.mrivanplays.ivancord.api.event;
+
+import com.mrivanplays.ivancord.api.Position;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+import net.md_5.bungee.api.connection.ProxiedPlayer;
+import net.md_5.bungee.api.plugin.Cancellable;
+import net.md_5.bungee.api.plugin.Event;
+
+/**
+ * Event, called when player is about to change his position.
+ */
+@ToString
+@EqualsAndHashCode(callSuper = false)
+@Getter
+public class PlayerChangePositionEvent extends Event implements Cancellable
+{
+
+    @Setter
+    private boolean cancelled;
+
+    /**
+     * The {@link ProxiedPlayer}, which changed position.
+     */
+    private ProxiedPlayer player;
+
+    /**
+     * The position the {@link ProxiedPlayer} moved from.
+     */
+    private Position movedFrom;
+
+    /**
+     * The position the {@link ProxiedPlayer} moved to.
+     */
+    @Setter
+    private Position movedTo;
+
+    public PlayerChangePositionEvent(ProxiedPlayer player, Position movedFrom, Position movedTo)
+    {
+        this.player = player;
+        this.movedFrom = movedFrom;
+        this.movedTo = movedTo;
+    }
+}
diff --git a/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java b/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java
index b14463bd6864054f38988d7bdaaa105635a3a263..03bc8c4b0cfa9733b93e74546a51bca1b15bca44 100644
--- a/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java
+++ b/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java
@@ -13,6 +13,7 @@ import net.md_5.bungee.api.chat.BaseComponent;
 import net.md_5.bungee.api.config.ServerInfo;
 import net.md_5.bungee.api.event.ServerConnectEvent;
 import net.md_5.bungee.api.score.Scoreboard;
+import com.mrivanplays.ivancord.api.Position; // IvanCord
 
 /**
  * Represents a player who's connection is being connected to somewhere else,
@@ -283,6 +284,16 @@ public interface ProxiedPlayer extends Connection, CommandSender
      */
     byte getViewDistance();
 
+    // IvanCord start - add player position api
+
+    /**
+     * Gets the player's position
+     *
+     * @return player position
+     */
+    Position getPosition();
+    // IvanCord end
+
     /**
      * Gets this player's chat mode.
      *
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/AbstractPacketHandler.java b/protocol/src/main/java/net/md_5/bungee/protocol/AbstractPacketHandler.java
index 0c334afcfc11a9893dfcbce8671d928569384e21..5606ba665f170d1c92f7a38c52cbe6253355e840 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/AbstractPacketHandler.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/AbstractPacketHandler.java
@@ -29,6 +29,7 @@ import net.md_5.bungee.protocol.packet.PlayerListHeaderFooter;
 import net.md_5.bungee.protocol.packet.PlayerListItem;
 import net.md_5.bungee.protocol.packet.PlayerListItemRemove;
 import net.md_5.bungee.protocol.packet.PlayerListItemUpdate;
+import net.md_5.bungee.protocol.packet.PlayerPosition; // IvanCord
 import net.md_5.bungee.protocol.packet.PluginMessage;
 import net.md_5.bungee.protocol.packet.Respawn;
 import net.md_5.bungee.protocol.packet.ScoreboardDisplay;
@@ -225,6 +226,11 @@ public abstract class AbstractPacketHandler
     public void handle(ServerData serverData) throws Exception
     {
     }
+    // IvanCord start
+    public void handle(PlayerPosition playerPosition) throws Exception
+    {
+    }
+    // IvanCord end
 
     // Waterfall start
     public void handle(EntityEffect entityEffect) throws Exception
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
index edb979c5a35ee292f011b481c18618df224a6e75..f4fb308e183ae5407771035298e60ff295bab379 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
@@ -37,6 +37,7 @@ import net.md_5.bungee.protocol.packet.PlayerListHeaderFooter;
 import net.md_5.bungee.protocol.packet.PlayerListItem;
 import net.md_5.bungee.protocol.packet.PlayerListItemRemove;
 import net.md_5.bungee.protocol.packet.PlayerListItemUpdate;
+import net.md_5.bungee.protocol.packet.PlayerPosition;
 import net.md_5.bungee.protocol.packet.PluginMessage;
 import net.md_5.bungee.protocol.packet.Respawn;
 import net.md_5.bungee.protocol.packet.ScoreboardDisplay;
@@ -480,6 +481,22 @@ public enum Protocol
                     map( ProtocolConstants.MINECRAFT_1_19_1, 0x0D ),
                     map( ProtocolConstants.MINECRAFT_1_19_3, 0x0C )
             );
+            // IvanCord start - add player position
+            TO_SERVER.registerPacket(
+                    PlayerPosition.class,
+                    PlayerPosition::new,
+                    map( ProtocolConstants.MINECRAFT_1_8, 0x06 ),
+                    map( ProtocolConstants.MINECRAFT_1_9, 0x0D ),
+                    map( ProtocolConstants.MINECRAFT_1_12, 0x0F ),
+                    map( ProtocolConstants.MINECRAFT_1_12_1, 0x0E ),
+                    map( ProtocolConstants.MINECRAFT_1_13, 0x11 ),
+                    map( ProtocolConstants.MINECRAFT_1_14, 0x12 ),
+                    map( ProtocolConstants.MINECRAFT_1_16, 0x13 ),
+                    map( ProtocolConstants.MINECRAFT_1_17, 0x12 ),
+                    map( ProtocolConstants.MINECRAFT_1_19, 0x14 ),
+                    map( ProtocolConstants.MINECRAFT_1_19_1, 0x15 )
+            );
+            // IvanCord end
         }
     },
     // 1
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/packet/PlayerPosition.java b/protocol/src/main/java/net/md_5/bungee/protocol/packet/PlayerPosition.java
new file mode 100644
index 0000000000000000000000000000000000000000..f23fe50e8c9a22ce934f5797f9e75fb2767a9479
--- /dev/null
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/packet/PlayerPosition.java
@@ -0,0 +1,54 @@
+// IvanCord start
+package net.md_5.bungee.protocol.packet;
+
+import io.netty.buffer.ByteBuf;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.NoArgsConstructor;
+import net.md_5.bungee.protocol.AbstractPacketHandler;
+import net.md_5.bungee.protocol.DefinedPacket;
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Data
+@EqualsAndHashCode(callSuper = false)
+public class PlayerPosition extends DefinedPacket
+{
+
+    private double x;
+    private double y;
+    private double z;
+    private float yaw;
+    private float pitch;
+    private boolean onGround;
+
+    @Override
+    public void read(ByteBuf buf)
+    {
+        x = buf.readDouble();
+        y = buf.readDouble();
+        z = buf.readDouble();
+        yaw = buf.readFloat();
+        pitch = buf.readFloat();
+        onGround = buf.readBoolean();
+    }
+
+    @Override
+    public void write(ByteBuf buf)
+    {
+        buf.writeDouble( x );
+        buf.writeDouble( y );
+        buf.writeDouble( z );
+        buf.writeFloat( yaw );
+        buf.writeFloat( pitch );
+        buf.writeBoolean( onGround );
+    }
+
+    @Override
+    public void handle(AbstractPacketHandler handler) throws Exception
+    {
+        handler.handle( this );
+    }
+}
+// IvanCord end
diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 15f855e0acce54bf4eda48a725c40aa75a97d9ae..2241ca9b5320bb433744212c5e494d9f827bfa83 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -66,6 +66,7 @@ import net.md_5.bungee.tab.ServerUnique;
 import net.md_5.bungee.tab.TabList;
 import net.md_5.bungee.util.CaseInsensitiveSet;
 import net.md_5.bungee.util.ChatComponentTransformer;
+import com.mrivanplays.ivancord.api.Position; // IvanCord
 
 @RequiredArgsConstructor
 public final class UserConnection implements ProxiedPlayer
@@ -106,6 +107,10 @@ public final class UserConnection implements ProxiedPlayer
     @Getter
     @Setter
     private int gamemode;
+    // IvanCord start
+    @Getter
+    private Position position;
+    // IvanCord end
     @Getter
     private int compressionThreshold = -1;
     // Used for trying multiple servers in order
@@ -669,6 +674,12 @@ public final class UserConnection implements ProxiedPlayer
     {
         return ( settings != null ) ? settings.getViewDistance() : 10;
     }
+    // IvanCord start
+    public void setPosition(Position position)
+    {
+        this.position = position;
+    }
+    // IvanCord end
 
     @Override
     public ProxiedPlayer.ChatMode getChatMode()
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index 652c202c9940fedf688e792400f4c0ab8a1de9a9..020872111781d12132c3c5fafc92454aa0490584 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -810,6 +810,39 @@ public class DownstreamBridge extends PacketHandler
         }
     }
 
+    // IvanCord start
+    @Override
+    public void handle(net.md_5.bungee.protocol.packet.PlayerPosition pac)
+    {
+        com.mrivanplays.ivancord.api.Position packetPos = new com.mrivanplays.ivancord.api.Position( pac.getX(), pac.getY(), pac.getZ(), pac.getYaw(), pac.getPitch(), pac.isOnGround() );
+        com.mrivanplays.ivancord.api.Position conPos = con.getPosition();
+        boolean changed = true;
+        if ( packetPos.equals( conPos ) )
+        {
+            changed = false;
+        }
+        if ( changed )
+        {
+            com.mrivanplays.ivancord.api.event.PlayerChangePositionEvent event = new com.mrivanplays.ivancord.api.event.PlayerChangePositionEvent( con, conPos, packetPos );
+            if ( event.callEvent() )
+            {
+                con.unsafe().sendPacket( pac );
+            } else
+            {
+                if ( !event.getMovedTo().equals( packetPos ) )
+                {
+                    com.mrivanplays.ivancord.api.Position pn = event.getMovedTo();
+                    con.unsafe().sendPacket( new net.md_5.bungee.protocol.packet.PlayerPosition( pn.getX(), pn.getY(), pn.getZ(), pn.getYaw(), pn.getPitch(), pn.isOnGround() ) );
+                } else
+                {
+                    con.unsafe().sendPacket( pac );
+                }
+            }
+            throw CancelSendSignal.INSTANCE;
+        }
+    }
+    // IvanCord end
+
     @Override
     public void handle(ServerData serverData) throws Exception
     {
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
index 71f05ea727712e5c3c5ba6e9e7d81826c5220a0e..19e19d24ae7ebdcce36a6186a991bb2d38ac361a 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
@@ -332,6 +332,14 @@ public class UpstreamBridge extends PacketHandler
         con.getPendingConnection().relayMessage( pluginMessage );
     }
 
+    // IvanCord start
+    @Override
+    public void handle(net.md_5.bungee.protocol.packet.PlayerPosition pac)
+    {
+        con.setPosition( new com.mrivanplays.ivancord.api.Position( pac.getX(), pac.getY(), pac.getZ(), pac.getYaw(), pac.getPitch(), pac.isOnGround() ) );
+    }
+    // IvanCord end
+
     @Override
     public String toString()
     {
