From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Thu, 13 Jun 2019 18:48:03 +0300
Subject: [PATCH] Add gplugins command


diff --git a/proxy/src/main/java/com/mrivanplays/ivancord/commands/CommandPlugins.java b/proxy/src/main/java/com/mrivanplays/ivancord/commands/CommandPlugins.java
new file mode 100644
index 0000000000000000000000000000000000000000..5d6a1d0c22a94b049636c40fbbe3e9de6b9704f1
--- /dev/null
+++ b/proxy/src/main/java/com/mrivanplays/ivancord/commands/CommandPlugins.java
@@ -0,0 +1,37 @@
+package com.mrivanplays.ivancord.commands;
+
+import java.util.Collection;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import net.md_5.bungee.api.ChatColor;
+import net.md_5.bungee.api.CommandSender;
+import net.md_5.bungee.api.ProxyServer;
+import net.md_5.bungee.api.chat.TextComponent;
+import net.md_5.bungee.api.plugin.Command;
+import net.md_5.bungee.api.plugin.Plugin;
+
+public class CommandPlugins extends Command
+{
+
+    public CommandPlugins()
+    {
+        super( "gplugins", "ivancord.command.plugins" );
+    }
+
+    @Override
+    public void execute(CommandSender sender, String[] args)
+    {
+        ProxyServer proxy = ProxyServer.getInstance();
+        Collection<Plugin> pluginsCol = proxy.getPluginManager().getPlugins();
+        StringBuilder pluginsList = new StringBuilder();
+        for ( Plugin plugin : pluginsCol )
+        {
+            pluginsList.append( ChatColor.GREEN.toString() ).append( plugin.getDescription().getName() )
+                .append( ChatColor.WHITE.toString() ).append( ", " );
+        }
+        String plugins = pluginsList.substring( 0, pluginsList.lastIndexOf( "," ) - 1 );
+        String message = proxy.getTranslation( "command_plugins_message", pluginsCol.size(), plugins );
+        sender.sendMessage( TextComponent.fromLegacyText( message ) );
+    }
+}
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index f9fa78c6b6c172d04dba9ac9111ba6de855eb617..aa870c1e400d4db3fd76088cabaedd1654a8077e 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -101,6 +101,7 @@ import com.mrivanplays.ivancord.modules.CommandFind;
 import com.mrivanplays.ivancord.modules.CommandList;
 import com.mrivanplays.ivancord.modules.CommandSend;
 import com.mrivanplays.ivancord.modules.CommandServer;
+import com.mrivanplays.ivancord.commands.CommandPlugins;
 // IvanCord end
 
 /**
@@ -250,6 +251,10 @@ public class BungeeCord extends ProxyServer
         getPluginManager().registerCommand( null, new CommandServer() );
         // IvanCord end
 
+        // IvanCord start - specific commands to IvanCord
+        getPluginManager().registerCommand( null, new CommandPlugins() );
+        // IvanCord end
+
         if ( !Boolean.getBoolean( "net.md_5.bungee.native.disable" ) )
         {
             if ( EncryptionUtil.nativeFactory.load() )
diff --git a/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java b/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
index 994bc589f26adb06ffe28313679ba6299878a6c8..51ee3ae3075cecaf63b7b624bc60450b36bf79c9 100644
--- a/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
+++ b/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
@@ -106,7 +106,8 @@ public class YamlConfig implements ConfigurationAdapter
             } ) );
             set( "permissions.admin", Arrays.asList( new String[]
             {
-                "bungeecord.command.alert", "bungeecord.command.end", "bungeecord.command.ip", "bungeecord.command.reload"
+                "bungeecord.command.alert", "bungeecord.command.end", "bungeecord.command.ip", "bungeecord.command.reload",
+                    "ivancord.command.plugins" // IvanCord - plugins command
             } ) );
         }
 
diff --git a/proxy/src/main/resources/messages.properties b/proxy/src/main/resources/messages.properties
index e838024539689e7ff1109d116434ddc64c9e088f..55493ee39edcbd362ae43d7c748cadf80c98e7ab 100644
--- a/proxy/src/main/resources/messages.properties
+++ b/proxy/src/main/resources/messages.properties
@@ -39,3 +39,4 @@ command_perms_groups=\u00a76You have the following groups: {0}
 command_perms_permission=\u00a79- {0}
 command_ip=\u00a79IP of {0} is {1}
 kick_to_lobby=You were kicked from server {0} with reason {1} . The fallback sent you to {2}
+command_plugins_message=Plugins ({0}): {1}
