From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Thu, 13 Jun 2019 18:03:14 +0300
Subject: [PATCH] Add option to disable tab list rewrite


diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
index 26467bd6c3519a80234a5c65e8208c073ac2f6ba..d6dd1125c33aff926dca2301c62ab6bb15156c2c 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
@@ -265,5 +265,13 @@ public interface ProxyConfig
      * @return should the player kicked be transferred to lobby
      */
     boolean isKickToLobby();
+
+    /**
+     * If this is true, the tab list rewriting will be disabled.
+     *
+     * @return should proxy rewrite TabList
+     */
+    boolean isDisableTabListRewrite();
+
     // IvanCord end
 }
diff --git a/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java b/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
index 238bc3c1f43aa5bd6de34f4355ede2dee818fef3..996c02b848c32e44abadb7ddeff9e59d57bc4449 100644
--- a/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
+++ b/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
@@ -11,6 +11,7 @@ public class IvanCordConfiguration extends WaterfallConfiguration
 {
 
     private boolean kickToLobby = true;
+    private boolean disableTabListRewrite = isOnlineMode();
 
     @Override
     public void load()
@@ -19,5 +20,6 @@ public class IvanCordConfiguration extends WaterfallConfiguration
         YamlConfig config = new YamlConfig( new File( "ivancord.yml" ) );
         config.load( false );
         kickToLobby = config.getBoolean( "kick_to_lobby", kickToLobby );
+        disableTabListRewrite = config.getBoolean( "disable_tablist_rewrite", disableTabListRewrite );
     }
 }
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index 92630789cbdcbe54f0f605d700508f74f4031e10..155bf842484dd3769e8ab000924b471498145c9f 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -174,8 +174,19 @@ public class DownstreamBridge extends PacketHandler
     @Override
     public void handle(PlayerListItem playerList) throws Exception
     {
+        // IvanCord start - rewrite the following code
+        /*
         con.getTabListHandler().onUpdate( TabList.rewrite( playerList ) );
         throw CancelSendSignal.INSTANCE; // Always throw because of profile rewriting
+         */
+        boolean skipRewrite = bungee.getConfig().isDisableTabListRewrite();
+        con.getTabListHandler().onUpdate( skipRewrite ? playerList : TabList.rewrite( playerList ) );
+        if ( !skipRewrite )
+        {
+            // only throw when we do rewriting
+            throw CancelSendSignal.INSTANCE;
+        }
+        // IvanCord end
     }
 
     // IvanCord start - remove these from handling
