From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Thu, 13 Jun 2019 18:03:14 +0300
Subject: [PATCH] Add option to disable tab list rewrite


diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
index 464018025984786e2d4ec322ea128314bd1a994e..39279c144ce7353682fd9340c3a6589025e82277 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
@@ -242,5 +242,13 @@ public interface ProxyConfig
      * @return should the player kicked be transferred to lobby
      */
     boolean isKickToLobby();
+
+    /**
+     * If this is true, the tab list rewriting will be disabled.
+     *
+     * @return should proxy rewrite TabList
+     */
+    boolean isDisableTabListRewrite();
+
     // IvanCord end
 }
diff --git a/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java b/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
index 238bc3c1f43aa5bd6de34f4355ede2dee818fef3..996c02b848c32e44abadb7ddeff9e59d57bc4449 100644
--- a/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
+++ b/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
@@ -11,6 +11,7 @@ public class IvanCordConfiguration extends WaterfallConfiguration
 {
 
     private boolean kickToLobby = true;
+    private boolean disableTabListRewrite = isOnlineMode();
 
     @Override
     public void load()
@@ -19,5 +20,6 @@ public class IvanCordConfiguration extends WaterfallConfiguration
         YamlConfig config = new YamlConfig( new File( "ivancord.yml" ) );
         config.load( false );
         kickToLobby = config.getBoolean( "kick_to_lobby", kickToLobby );
+        disableTabListRewrite = config.getBoolean( "disable_tablist_rewrite", disableTabListRewrite );
     }
 }
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index f7ec4d4f6fc9fb2c3a0cd3723b48faaedd1d9492..1b9172932520c11371cbda259b519e46224a7431 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -165,8 +165,19 @@ public class DownstreamBridge extends PacketHandler
     @Override
     public void handle(PlayerListItem playerList) throws Exception
     {
+        // IvanCord start - rewrite the following code
+        /*
         con.getTabListHandler().onUpdate( TabList.rewrite( playerList ) );
         throw CancelSendSignal.INSTANCE; // Always throw because of profile rewriting
+         */
+        boolean skipRewrite = bungee.getConfig().isDisableTabListRewrite();
+        con.getTabListHandler().onUpdate( skipRewrite ? playerList : TabList.rewrite( playerList ) );
+        if ( !skipRewrite )
+        {
+            // only throw when we do rewriting
+            throw CancelSendSignal.INSTANCE;
+        }
+        // IvanCord end
     }
 
     // IvanCord start - remove these from handling
