From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Tue, 25 Jun 2019 16:03:22 +0300
Subject: [PATCH] Add LoginCancelledEvent


diff --git a/api/src/main/java/com/mrivanplays/ivancord/api/event/LoginCancelledEvent.java b/api/src/main/java/com/mrivanplays/ivancord/api/event/LoginCancelledEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..f140b0195e9661e364dafab7c3aba7ec654d8a52
--- /dev/null
+++ b/api/src/main/java/com/mrivanplays/ivancord/api/event/LoginCancelledEvent.java
@@ -0,0 +1,54 @@
+package com.mrivanplays.ivancord.api.event;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import net.md_5.bungee.api.connection.PendingConnection;
+import net.md_5.bungee.api.plugin.Event;
+
+/**
+ * Represents a event, called when login process is aborted/cancelled
+ * by the logging player or the proxy. In most of the times, the event
+ * is being triggered by the logging player pressing the "Cancel"
+ * button on the client.
+ */
+@Getter
+@AllArgsConstructor
+public class LoginCancelledEvent extends Event
+{
+
+    /**
+     * The pending connection, which aborted their login process
+     */
+    private PendingConnection connection;
+
+    /**
+     * The reason of which the event was being triggered
+     */
+    private CancelReason cancelReason;
+
+    /**
+     * Represents a cancel reason
+     */
+    public static enum CancelReason
+    {
+
+        /**
+         * The client aborted the login process by the user
+         * pressing the "Cancel" button
+         */
+        CLIENT,
+
+        /**
+         * {@link net.md_5.bungee.api.event.LoginEvent} or
+         * {@link net.md_5.bungee.api.event.PreLoginEvent} are getting cancelled
+         */
+        PROXY,
+
+        /**
+         * Exception were caught during login process and the
+         * proxy cancelled the login process
+         */
+        PROXY_EXCEPTION
+        ;
+    }
+}
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index 0fd58e83a47b46e218d6bf49e4f5f27e6761a77a..b7e37d47a4c4476b6c1d1b604db05835981d324a 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -68,6 +68,7 @@ import net.md_5.bungee.util.AllowedCharacters;
 import net.md_5.bungee.util.BoundedArrayList;
 import net.md_5.bungee.util.BufUtil;
 import net.md_5.bungee.util.QuietException;
+import com.mrivanplays.ivancord.api.event.LoginCancelledEvent; // IvanCord
 
 @RequiredArgsConstructor
 public class InitialHandler extends PacketHandler implements PendingConnection
@@ -144,6 +145,9 @@ public class InitialHandler extends PacketHandler implements PendingConnection
         {
             ch.close();
         }
+        // IvanCord start - implement LoginCancelledEvent
+        new LoginCancelledEvent( InitialHandler.this, LoginCancelledEvent.CancelReason.PROXY_EXCEPTION ).callEvent();
+        // IvanCord end
     }
 
     @Override
@@ -395,11 +399,17 @@ public class InitialHandler extends PacketHandler implements PendingConnection
             {
                 if ( result.isCancelled() )
                 {
+                    // IvanCord start - implement LoginCancelledEvent
+                    new LoginCancelledEvent( InitialHandler.this, LoginCancelledEvent.CancelReason.PROXY ).callEvent();
+                    // IvanCord end
                     disconnect( result.getCancelReasonComponents() );
                     return;
                 }
                 if ( ch.isClosed() )
                 {
+                    // IvanCord start - implement LoginCancelledEvent
+                    new LoginCancelledEvent( InitialHandler.this, LoginCancelledEvent.CancelReason.CLIENT ).callEvent();
+                    // IvanCord end
                     return;
                 }
                 if ( onlineMode )
@@ -525,11 +535,17 @@ public class InitialHandler extends PacketHandler implements PendingConnection
             {
                 if ( result.isCancelled() )
                 {
+                    // IvanCord start - implement login cancelled event
+                    new LoginCancelledEvent( InitialHandler.this, LoginCancelledEvent.CancelReason.PROXY ).callEvent();
+                    // IvanCord end
                     disconnect( result.getCancelReasonComponents() );
                     return;
                 }
                 if ( ch.isClosed() )
                 {
+                    // IvanCord start - implement LoginCancelledEvent
+                    new LoginCancelledEvent( InitialHandler.this, LoginCancelledEvent.CancelReason.CLIENT ).callEvent();
+                    // IvanCord end
                     return;
                 }
 
@@ -564,6 +580,12 @@ public class InitialHandler extends PacketHandler implements PendingConnection
 
                             userCon.connect( server, null, true, ServerConnectEvent.Reason.JOIN_PROXY );
                         }
+                        // IvanCord start
+                        else
+                        {
+                            new LoginCancelledEvent( InitialHandler.this, LoginCancelledEvent.CancelReason.CLIENT ).callEvent();
+                        }
+                        // IvanCord end
                     }
                 } );
             }
