From dfbbf5abf65c7e13d1842412c9e0ce9985a88659 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Thu, 13 Jun 2019 18:48:03 +0300
Subject: [PATCH] Add gplugins command


diff --git a/proxy/src/main/java/com/github/mrivanplays/ivancord/commands/CommandPlugins.java b/proxy/src/main/java/com/github/mrivanplays/ivancord/commands/CommandPlugins.java
new file mode 100644
index 00000000..2673cce7
--- /dev/null
+++ b/proxy/src/main/java/com/github/mrivanplays/ivancord/commands/CommandPlugins.java
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2019 Ivan Pekov (MrIvanPlays)
+
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in the
+ * Software without restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
+ * and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+
+ * The above copyright notice and this permission notice shall be included in all copies
+ * or substantial portions of the Software.
+
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ **/
+package com.github.mrivanplays.ivancord.commands;
+
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import net.md_5.bungee.Util;
+import net.md_5.bungee.api.CommandSender;
+import net.md_5.bungee.api.ProxyServer;
+import net.md_5.bungee.api.chat.TextComponent;
+import net.md_5.bungee.api.plugin.Command;
+
+public class CommandPlugins extends Command
+{
+
+    public CommandPlugins()
+    {
+        super( "gplugins", "ivancord.command.plugins" );
+    }
+
+    @Override
+    public void execute(CommandSender sender, String[] args)
+    {
+        ProxyServer proxy = ProxyServer.getInstance();
+        Set<String> pluginNames = proxy.getPluginManager().getPlugins()
+                .stream()
+                .map( plugin -> plugin.getDescription().getName() )
+                .collect( Collectors.toSet() );
+        String message = proxy.getTranslation( "command_plugins_message", pluginNames.size(), Util.csv( pluginNames ) );
+        sender.sendMessage( TextComponent.fromLegacyText( message ) );
+    }
+}
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 8ee54d80..f2fdbf54 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -91,6 +91,7 @@ import net.md_5.bungee.query.RemoteQuery;
 import net.md_5.bungee.scheduler.BungeeScheduler;
 import net.md_5.bungee.util.CaseInsensitiveMap;
 
+import com.github.mrivanplays.ivancord.commands.CommandPlugins;
 import com.github.mrivanplays.ivancord.modules.CommandAlert;
 import com.github.mrivanplays.ivancord.modules.CommandAlertRaw;
 import com.github.mrivanplays.ivancord.modules.CommandFind;
@@ -241,6 +242,10 @@ public class BungeeCord extends ProxyServer
         getPluginManager().registerCommand( null, new CommandServer() );
         // IvanCord end
 
+        // IvanCord start - specific commands to IvanCord
+        getPluginManager().registerCommand( null, new CommandPlugins() );
+        // IvanCord end
+
         if ( !Boolean.getBoolean( "net.md_5.bungee.native.disable" ) )
         {
             if ( EncryptionUtil.nativeFactory.load() )
diff --git a/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java b/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
index 9c187f73..5021af15 100644
--- a/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
+++ b/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
@@ -103,7 +103,8 @@ public class YamlConfig implements ConfigurationAdapter
             } ) );
             set( "permissions.admin", Arrays.asList( new String[]
             {
-                "bungeecord.command.alert", "bungeecord.command.end", "bungeecord.command.ip", "bungeecord.command.reload"
+                "bungeecord.command.alert", "bungeecord.command.end", "bungeecord.command.ip", "bungeecord.command.reload",
+                    "ivancord.command.plugins" // IvanCord - plugins command
             } ) );
         }
 
diff --git a/proxy/src/main/resources/messages.properties b/proxy/src/main/resources/messages.properties
index d917c74a..b85b5163 100644
--- a/proxy/src/main/resources/messages.properties
+++ b/proxy/src/main/resources/messages.properties
@@ -25,3 +25,4 @@ name_invalid=Username contains invalid characters.
 ping_cannot_connect=\u00a7c[Bungee] Can't connect to server.
 offline_mode_player=Not authenticated with Minecraft.net
 kick_to_lobby=You were kicked from server {0} with reason {1}
+command_plugins_message=Plugins({0}): \u00a7a{1}
-- 
2.21.0.windows.1

