From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Tue, 25 Jun 2019 16:03:22 +0300
Subject: [PATCH] Add LoginCancelledEvent


diff --git a/api/src/main/java/com/mrivanplays/ivancord/api/event/LoginCancelledEvent.java b/api/src/main/java/com/mrivanplays/ivancord/api/event/LoginCancelledEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..f140b0195e9661e364dafab7c3aba7ec654d8a52
--- /dev/null
+++ b/api/src/main/java/com/mrivanplays/ivancord/api/event/LoginCancelledEvent.java
@@ -0,0 +1,54 @@
+package com.mrivanplays.ivancord.api.event;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import net.md_5.bungee.api.connection.PendingConnection;
+import net.md_5.bungee.api.plugin.Event;
+
+/**
+ * Represents a event, called when login process is aborted/cancelled
+ * by the logging player or the proxy. In most of the times, the event
+ * is being triggered by the logging player pressing the "Cancel"
+ * button on the client.
+ */
+@Getter
+@AllArgsConstructor
+public class LoginCancelledEvent extends Event
+{
+
+    /**
+     * The pending connection, which aborted their login process
+     */
+    private PendingConnection connection;
+
+    /**
+     * The reason of which the event was being triggered
+     */
+    private CancelReason cancelReason;
+
+    /**
+     * Represents a cancel reason
+     */
+    public static enum CancelReason
+    {
+
+        /**
+         * The client aborted the login process by the user
+         * pressing the "Cancel" button
+         */
+        CLIENT,
+
+        /**
+         * {@link net.md_5.bungee.api.event.LoginEvent} or
+         * {@link net.md_5.bungee.api.event.PreLoginEvent} are getting cancelled
+         */
+        PROXY,
+
+        /**
+         * Exception were caught during login process and the
+         * proxy cancelled the login process
+         */
+        PROXY_EXCEPTION
+        ;
+    }
+}
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index 37ea35e3021faff572b2a2f2b562bc36a8fb975e..ed525efadd950d34a0f6638117facedc81b76143 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -65,7 +65,7 @@ import net.md_5.bungee.protocol.packet.StatusResponse;
 import net.md_5.bungee.util.BoundedArrayList;
 import net.md_5.bungee.util.BufUtil;
 import net.md_5.bungee.util.QuietException;
-
+import com.mrivanplays.ivancord.api.event.LoginCancelledEvent; // IvanCord
 @RequiredArgsConstructor
 public class InitialHandler extends PacketHandler implements PendingConnection
 {
@@ -139,6 +139,9 @@ public class InitialHandler extends PacketHandler implements PendingConnection
         {
             ch.close();
         }
+        // IvanCord start - implement LoginCancelledEvent
+        new LoginCancelledEvent( InitialHandler.this, LoginCancelledEvent.CancelReason.PROXY_EXCEPTION ).callEvent();
+        // IvanCord end
     }
 
     @Override
@@ -396,11 +399,17 @@ public class InitialHandler extends PacketHandler implements PendingConnection
             {
                 if ( result.isCancelled() )
                 {
+                    // IvanCord start - implement LoginCancelledEvent
+                    new LoginCancelledEvent( InitialHandler.this, LoginCancelledEvent.CancelReason.PROXY ).callEvent();
+                    // IvanCord end
                     disconnect( result.getCancelReasonComponents() );
                     return;
                 }
                 if ( ch.isClosed() )
                 {
+                    // IvanCord start - implement LoginCancelledEvent
+                    new LoginCancelledEvent( InitialHandler.this, LoginCancelledEvent.CancelReason.CLIENT ).callEvent();
+                    // IvanCord end
                     return;
                 }
                 if ( onlineMode )
@@ -517,11 +526,17 @@ public class InitialHandler extends PacketHandler implements PendingConnection
             {
                 if ( result.isCancelled() )
                 {
+                    // IvanCord start - implement login cancelled event
+                    new LoginCancelledEvent( InitialHandler.this, LoginCancelledEvent.CancelReason.PROXY ).callEvent();
+                    // IvanCord end
                     disconnect( result.getCancelReasonComponents() );
                     return;
                 }
                 if ( ch.isClosed() )
                 {
+                    // IvanCord start - implement LoginCancelledEvent
+                    new LoginCancelledEvent( InitialHandler.this, LoginCancelledEvent.CancelReason.CLIENT ).callEvent();
+                    // IvanCord end
                     return;
                 }
 
@@ -530,6 +545,8 @@ public class InitialHandler extends PacketHandler implements PendingConnection
                     @Override
                     public void run()
                     {
+                        // IvanCord start - recreate this in favour of LoginCancelledEvent
+                        /*
                         if ( !ch.isClosing() )
                         {
                             UserConnection userCon = new UserConnection( bungee, ch, getName(), InitialHandler.this );
@@ -558,6 +575,38 @@ public class InitialHandler extends PacketHandler implements PendingConnection
 
                             thisState = State.FINISHED;
                         }
+                         */
+                        if ( ch.isClosing() )
+                        {
+                            new LoginCancelledEvent( InitialHandler.this, LoginCancelledEvent.CancelReason.CLIENT ).callEvent();
+                            return;
+                        }
+                        UserConnection userCon = new UserConnection( bungee, ch, getName(), InitialHandler.this );
+                        userCon.setCompressionThreshold( BungeeCord.getInstance().config.getCompressionThreshold() );
+                        userCon.init();
+
+                        unsafe.sendPacket( new LoginSuccess( getUniqueId().toString(), getName() ) ); // With dashes in between
+                        ch.setProtocol( Protocol.GAME );
+
+                        ch.getHandle().pipeline().get( HandlerBoss.class ).setHandler( new UpstreamBridge( bungee, userCon ) );
+                        bungee.getPluginManager().callEvent( new PostLoginEvent( userCon ) );
+                        ServerInfo server;
+                        if ( bungee.getReconnectHandler() != null )
+                        {
+                            server = bungee.getReconnectHandler().getServer( userCon );
+                        } else
+                        {
+                            server = AbstractReconnectHandler.getForcedHost( InitialHandler.this );
+                        }
+                        if ( server == null )
+                        {
+                            server = bungee.getServerInfo( listener.getDefaultServer() );
+                        }
+
+                        userCon.connect( server, null, true, ServerConnectEvent.Reason.JOIN_PROXY );
+
+                        thisState = State.FINISHED;
+                        // IvanCord end
                     }
                 } );
             }
