From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Thu, 27 Jun 2019 18:25:10 +0300
Subject: [PATCH] Add plugin disable api


diff --git a/api/src/main/java/com/mrivanplays/ivancord/api/event/PluginDisabledEvent.java b/api/src/main/java/com/mrivanplays/ivancord/api/event/PluginDisabledEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..6e3335e07bd197ed17c52a8018c732a92748d2fd
--- /dev/null
+++ b/api/src/main/java/com/mrivanplays/ivancord/api/event/PluginDisabledEvent.java
@@ -0,0 +1,20 @@
+package com.mrivanplays.ivancord.api.event;
+
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import net.md_5.bungee.api.plugin.Event;
+import net.md_5.bungee.api.plugin.Plugin;
+
+/**
+ * Represents a event, called when a plugin was disabled
+ */
+@Data
+@EqualsAndHashCode(callSuper = false)
+public class PluginDisabledEvent extends Event
+{
+
+    /**
+     * The plugin which was disabled
+     */
+    private final Plugin plugin;
+}
diff --git a/api/src/main/java/com/mrivanplays/ivancord/api/event/PluginEnabledEvent.java b/api/src/main/java/com/mrivanplays/ivancord/api/event/PluginEnabledEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..ab7c3978a708ff47c48dc56804cbdd753d870852
--- /dev/null
+++ b/api/src/main/java/com/mrivanplays/ivancord/api/event/PluginEnabledEvent.java
@@ -0,0 +1,20 @@
+package com.mrivanplays.ivancord.api.event;
+
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import net.md_5.bungee.api.plugin.Event;
+import net.md_5.bungee.api.plugin.Plugin;
+
+/**
+ * Represents a event, called when a plugin was enabled
+ */
+@Data
+@EqualsAndHashCode(callSuper = false)
+public class PluginEnabledEvent extends Event
+{
+
+    /**
+     * The plugin which was enabled.
+     */
+    private final Plugin plugin;
+}
diff --git a/api/src/main/java/net/md_5/bungee/api/plugin/Plugin.java b/api/src/main/java/net/md_5/bungee/api/plugin/Plugin.java
index 35556d48b70930252e545db0a890301b99a5421c..ba1890d63cdfa41e7db009e553698c33e4d98425 100644
--- a/api/src/main/java/net/md_5/bungee/api/plugin/Plugin.java
+++ b/api/src/main/java/net/md_5/bungee/api/plugin/Plugin.java
@@ -166,6 +166,19 @@ public class Plugin
         return getClass().getClassLoader().getResourceAsStream( name );
     }
 
+    // IvanCord start
+
+    /**
+     * Returns whenever this plugin is enabled.
+     *
+     * @return <code>true</code> if enabled
+     */
+    public boolean isEnabled()
+    {
+        return proxy.getPluginManager().isPluginEnabled( this );
+    }
+    // IvanCord end
+
     /**
      * Called by the loader to initialize the fields in this plugin.
      *
diff --git a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
index 4312e5714eb8f989e77c4e7447430c02eb392eaa..eea22ab19840860720abb36be76627242601cf58 100644
--- a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
+++ b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
@@ -23,7 +23,12 @@ import java.util.Set;
 import java.util.Stack;
 import java.util.jar.JarEntry;
 import java.util.jar.JarFile;
+import java.util.logging.Handler; // IvanCord
 import java.util.logging.Level;
+// IvanCord start
+import com.mrivanplays.ivancord.api.event.PluginDisabledEvent;
+import com.mrivanplays.ivancord.api.event.PluginEnabledEvent;
+// IvanCord end
 import lombok.RequiredArgsConstructor;
 import net.md_5.bungee.api.ChatColor;
 import net.md_5.bungee.api.CommandSender;
@@ -285,6 +290,9 @@ public class PluginManager
                 {
                     plugin.getDescription().getName(), plugin.getDescription().getVersion(), plugin.getDescription().getAuthor()
                 } );
+                // IvanCord start
+                callEvent( new PluginEnabledEvent( plugin ) );
+                // IvanCord end
             } catch ( Throwable t )
             {
                 ProxyServer.getInstance().getLogger().log( Level.WARNING, "Exception encountered when loading plugin: " + plugin.getDescription().getName(), t );
@@ -499,4 +507,78 @@ public class PluginManager
     {
         return Collections.unmodifiableCollection( commandMap.entrySet() );
     }
+
+    // IvanCord start - plugin disable api
+
+    /**
+     * @param pluginName the plugin's name you want to disable
+     * @throws IllegalArgumentException if plugin name is null/empty || found plugin is null
+     * @see #disablePlugin(Plugin) 
+     */
+    public void disablePlugin(String pluginName)
+    {
+        Preconditions.checkArgument( pluginName != null || !pluginName.isEmpty(), "Plugin name cannot be null/empty" );
+        Preconditions.checkArgument( getPlugin( pluginName ) != null, "Cannot find plugin for disable with name '" + pluginName + "'" );
+        disablePlugin( getPlugin( pluginName ) );
+    }
+
+    /**
+     * Disables the specified plugin, if it is enabled.
+     *
+     * @param plugin the plugin you want to disable
+     * @throws IllegalArgumentException if the plugin is null
+     */
+    public void disablePlugin(Plugin plugin)
+    {
+        Preconditions.checkNotNull( plugin, "Disabling plugin cannot be null" );
+        if ( !plugin.isEnabled() )
+        {
+            return;
+        }
+        try
+        {
+            plugin.onDisable();
+            for ( Handler handler : plugin.getLogger().getHandlers() )
+            {
+                handler.close();
+            }
+        } catch ( Throwable e )
+        {
+            proxy.getLogger().log( Level.SEVERE, "Exception disabling plugin " + plugin.getDescription().getName(), e );
+        }
+        unregisterListeners( plugin );
+        unregisterCommands( plugin );
+        proxy.getScheduler().cancel( plugin );
+        plugin.getExecutorService().shutdownNow();
+        plugins.remove( plugin.getDescription().getName(), plugin );
+        callEvent( new PluginDisabledEvent( plugin ) );
+        System.gc();
+    }
+
+    /**
+     * @param pluginName the plugin's name you want to check
+     * @return <code>true</code> if enabled, otherwise <code>false</code>
+     * @throws IllegalArgumentException if plugin name is null/empty || found plugin is null
+     * @see #isPluginEnabled(Plugin)
+     */
+    public boolean isPluginEnabled(String pluginName)
+    {
+        Preconditions.checkArgument( pluginName != null || !pluginName.isEmpty(), "Plugin name cannot be null/empty" );
+        Preconditions.checkArgument( getPlugin( pluginName ) != null, "Cannot find plugin to check with name '" + pluginName + "'" );
+        return isPluginEnabled( getPlugin( pluginName ) );
+    }
+
+    /**
+     * Returns whenever a plugin is enabled
+     *
+     * @param plugin plugin
+     * @return <code>true</code> if enabled, otherwise <code>false</code>
+     * @throws IllegalArgumentException if plugin is null
+     */
+    public boolean isPluginEnabled(Plugin plugin)
+    {
+        Preconditions.checkNotNull( plugin, "Checked plugin cannot be null" );
+        return getPlugins().contains( plugin );
+    }
+    // IvanCord end
 }
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 2e1ed41ab35994f954be90415c4c6c863368bb3e..ff568fc1cac5bc3520f1ffa5fd9e0cd646ed4cb2 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -44,6 +44,7 @@ import java.util.concurrent.locks.ReentrantReadWriteLock;
 import java.util.logging.Handler;
 import java.util.logging.Level;
 import java.util.logging.Logger;
+
 import lombok.Getter;
 import lombok.Setter;
 import lombok.Synchronized;
@@ -487,24 +488,13 @@ public class BungeeCord extends ProxyServer
                 saveThread.cancel();
                 // metricsThread.cancel(); // Waterfall: Disable Metrics
 
-                // TODO: Fix this shit
                 getLogger().info( "Disabling plugins" );
+                // IvanCord start - fix this shit
                 for ( Plugin plugin : Lists.reverse( new ArrayList<>( pluginManager.getPlugins() ) ) )
                 {
-                    try
-                    {
-                        plugin.onDisable();
-                        for ( Handler handler : plugin.getLogger().getHandlers() )
-                        {
-                            handler.close();
-                        }
-                    } catch ( Throwable t )
-                    {
-                        getLogger().log( Level.SEVERE, "Exception disabling plugin " + plugin.getDescription().getName(), t );
-                    }
-                    getScheduler().cancel( plugin );
-                    plugin.getExecutorService().shutdownNow();
+                    pluginManager.disablePlugin( plugin );
                 }
+                // IvanCord end
 
                 getLogger().info( "Closing IO threads" );
                 // IvanCord start - speed up shutdown
