From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Sun, 30 Jun 2019 21:13:31 +0300
Subject: [PATCH] Remove some of the TODOs


diff --git a/api/src/main/java/io/github/waterfallmc/waterfall/QueryResult.java b/api/src/main/java/io/github/waterfallmc/waterfall/QueryResult.java
index 9a262c8a11caeade95a084e5e5bc0be398917bbf..266a46e3abf3c4948477cb15c19a7e3b081bac02 100644
--- a/api/src/main/java/io/github/waterfallmc/waterfall/QueryResult.java
+++ b/api/src/main/java/io/github/waterfallmc/waterfall/QueryResult.java
@@ -56,4 +56,23 @@ public class QueryResult {
      */
     @NonNull
     private String version;
+    // IvanCord start
+    /**
+     * The plugins to return for this server
+     */
+    private final List<PluginInfo> plugins = new java.util.LinkedList<>();
+
+    /**
+     * Represents a PluginInfo, containing a name and version of a plugin
+     */
+    @Data
+    public static final class PluginInfo
+    {
+        @NonNull
+        private final String name;
+
+        @NonNull
+        private final String version;
+    }
+    // IvanCord end
 }
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeServerInfo.java b/proxy/src/main/java/net/md_5/bungee/BungeeServerInfo.java
index 7b002089ef9dd9f365dc0f7db076976b8dc6260c..1ab5a0b3c175a39ad8ca6a39ce2568be1345fc83 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeServerInfo.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeServerInfo.java
@@ -1,6 +1,7 @@
 package net.md_5.bungee;
 
 import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterables;
 import io.netty.bootstrap.Bootstrap;
 import io.netty.channel.ChannelFuture;
 import io.netty.channel.ChannelFutureListener;
@@ -103,13 +104,16 @@ public class BungeeServerInfo implements ServerInfo
         sendData( channel, data, true );
     }
 
-    // TODO: Don't like this method
+    // IvanCord start - remove to do "Don't like this method"
+    // We're rewriting it fully
     @Override
     public boolean sendData(String channel, byte[] data, boolean queue)
     {
         Preconditions.checkNotNull( channel, "channel" );
         Preconditions.checkNotNull( data, "data" );
 
+        // old code
+        /*
         Server server;
 
         synchronized (players) {
@@ -124,6 +128,24 @@ public class BungeeServerInfo implements ServerInfo
                 packetQueue.add(new PluginMessage(channel, data, false));
             }
         }
+         */
+
+        // here's ours
+        if ( players.isEmpty() )
+        {
+            if ( queue )
+            {
+                synchronized ( packetQueue )
+                {
+                    packetQueue.add( new PluginMessage( channel, data, false ) );
+                }
+            }
+        } else
+        {
+            Server server = Iterables.getFirst( players, null ).getServer();
+            server.sendData( channel, data );
+            return true;
+        }
 
         return false;
     }
diff --git a/proxy/src/main/java/net/md_5/bungee/ServerConnector.java b/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
index 21b56df8b9392cd63846060dc3576ede2da02643..f0311fb938450b636fdff263d4a7af10a8383226 100644
--- a/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
+++ b/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
@@ -387,7 +387,10 @@ public class ServerConnector extends PacketHandler
         }
 
         // Add to new server
-        // TODO: Move this to the connected() method of DownstreamBridge
+        // IvanCord - remove the to do "Move this to the connected() method of DownstreamBridge"
+        // the to do was removed because if you move this where it is said, the dimension change
+        // and the other stuff won't happen. Because of no dimension change set to false the
+        // connection will always get disconnected from the proxy if cant connect to the server he want.
         target.addPlayer( user );
         user.getPendingConnects().remove( target );
         user.setServerJoinQueue( null );
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
index f618468a521672e7716c035b3d311bdd821e1007..11d9f15d65223251a759e23c664182adb0ea7728 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
@@ -34,6 +34,7 @@ import net.md_5.bungee.protocol.packet.PluginMessage;
 import net.md_5.bungee.protocol.packet.TabCompleteRequest;
 import net.md_5.bungee.protocol.packet.TabCompleteResponse;
 import com.mrivanplays.ivancord.api.Position; // IvanCord
+import net.md_5.bungee.tab.ServerUnique; // IvanCord
 
 public class UpstreamBridge extends PacketHandler
 {
@@ -73,9 +74,9 @@ public class UpstreamBridge extends PacketHandler
             // Manually remove from everyone's tab list
             // since the packet from the server arrives
             // too late
-            // TODO: This should only done with server_unique
-            //       tab list (which is the only one supported
-            //       currently)
+            // IvanCord start - remove to do "This should only done with server_unique tab list"
+            if ( con.getTabListHandler() instanceof ServerUnique )
+            {
             PlayerListItem packet = new PlayerListItem();
             packet.setAction( PlayerListItem.Action.REMOVE_PLAYER );
             PlayerListItem.Item item = new PlayerListItem.Item();
@@ -88,6 +89,7 @@ public class UpstreamBridge extends PacketHandler
             {
                 player.unsafe().sendPacket( packet );
             }
+            } // IvanCord end
             con.getServer().disconnect( "Quitting" );
         }
     }
diff --git a/query/src/main/java/net/md_5/bungee/query/QueryHandler.java b/query/src/main/java/net/md_5/bungee/query/QueryHandler.java
index 1b9f054bc4a7b87ac3531e4e49d5e93d15420d26..03c18c38edbe4f25d770c8405d2924a06dd553c6 100644
--- a/query/src/main/java/net/md_5/bungee/query/QueryHandler.java
+++ b/query/src/main/java/net/md_5/bungee/query/QueryHandler.java
@@ -10,6 +10,7 @@ import io.netty.channel.ChannelHandlerContext;
 import io.netty.channel.SimpleChannelInboundHandler;
 import io.netty.channel.socket.DatagramPacket;
 import java.net.InetAddress;
+import java.util.Iterator; // IvanCord
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
@@ -22,6 +23,7 @@ import lombok.RequiredArgsConstructor;
 import net.md_5.bungee.api.ProxyServer;
 import net.md_5.bungee.api.config.ListenerInfo;
 import net.md_5.bungee.api.connection.ProxiedPlayer;
+import net.md_5.bungee.api.plugin.PluginDescription;
 
 @RequiredArgsConstructor
 public class QueryHandler extends SimpleChannelInboundHandler<DatagramPacket>
@@ -33,6 +35,7 @@ public class QueryHandler extends SimpleChannelInboundHandler<DatagramPacket>
     private final Random random = new Random();
     private final Cache<InetAddress, QuerySession> sessions = CacheBuilder.newBuilder().expireAfterWrite( 30, TimeUnit.SECONDS ).build();
     private static io.github.waterfallmc.waterfall.utils.FastException cachedNoSessionException = new io.github.waterfallmc.waterfall.utils.FastException("No Session!");
+    private volatile List<QueryResult.PluginInfo> cachedPluginInfoList = null; // IvanCord
 
     private void writeShort(ByteBuf buf, int s)
     {
@@ -99,13 +102,33 @@ public class QueryHandler extends SimpleChannelInboundHandler<DatagramPacket>
             {
                 throw cachedNoSessionException; // Waterfall
             }
-
+            // IvanCord start
+            List<QueryResult.PluginInfo> plugins = this.cachedPluginInfoList;
+            if ( plugins == null )
+            {
+                synchronized ( this )
+                {
+                    if ( cachedPluginInfoList == null )
+                    {
+                        cachedPluginInfoList = plugins = bungee.getPluginManager().getPlugins()
+                                .parallelStream()
+                                .map( plugin ->
+                                {
+                                    PluginDescription pluginDescription = plugin.getDescription();
+                                    return new QueryResult.PluginInfo( pluginDescription.getName(), pluginDescription.getVersion() );
+                                } )
+                                .collect( Collectors.toList() );
+                    }
+                }
+            }
+            // IvanCord end
             // Waterfall start
             List<String> players = bungee.getPlayers().stream().map(ProxiedPlayer::getName).collect(Collectors.toList());
 
             ProxyQueryEvent event = new ProxyQueryEvent(listener, new QueryResult(listener.getMotd(), "SMP", "IvanCord_Proxy", // IvanCord - rebranding
                     bungee.getOnlineCount(), listener.getMaxPlayers(), listener.getHost().getPort(),
                     listener.getHost().getHostString(), "MINECRAFT",  players, bungee.getGameVersion()));
+            event.getResult().getPlugins().addAll( plugins ); // IvanCord
             QueryResult result = bungee.getPluginManager().callEvent(event).getResult();
             // Waterfall end
 
@@ -139,7 +162,24 @@ public class QueryHandler extends SimpleChannelInboundHandler<DatagramPacket>
                 // Start Extra Info
                 data.put( "game_id", result.getGameId() );
                 data.put( "version", result.getVersion() );
-                data.put( "plugins", "" ); // TODO: Allow population?
+                // IvanCord start - remove to do "Allow population?"
+                /*
+                data.put( "plugins", "" );
+                 */
+                StringBuilder pluginsString = new StringBuilder();
+                pluginsString.append( bungee.getName() ).append( ' ' ).append( bungee.getGameVersion() ).append( ':' ).append( ' ' );
+                Iterator<QueryResult.PluginInfo> iterator = plugins.iterator();
+                while ( iterator.hasNext() )
+                {
+                    QueryResult.PluginInfo pluginInfo = iterator.next();
+                    pluginsString.append( pluginInfo.getName() ).append( ' ' ).append( pluginInfo.getVersion() );
+                    if ( iterator.hasNext() )
+                    {
+                        pluginsString.append( ';' ).append( ' ' );
+                    }
+                }
+                data.put( "plugins", pluginsString.toString() );
+                // IvanCord end
                 // End Extra Info
                 data.put( "map", result.getWorldName() );
                 data.put( "numplayers", Integer.toString( result.getOnlinePlayers() ) );
