From d3a16297cace9243b1ff712cb60524872c46bb5a Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Mon, 8 Jul 2019 15:50:25 +0300
Subject: [PATCH] Fix force_default_server not working


diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index 512dd1325..0213b019b 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -576,16 +576,18 @@ public class InitialHandler extends PacketHandler implements PendingConnection
                         ch.getHandle().pipeline().get( HandlerBoss.class ).setHandler( new UpstreamBridge( bungee, userCon ) );
                         bungee.getPluginManager().callEvent( new PostLoginEvent( userCon ) );
                         ServerInfo server;
-                        if ( bungee.getReconnectHandler() != null )
+                        if ( listener.isForceDefault() )
                         {
-                            server = bungee.getReconnectHandler().getServer( userCon );
+                            server = bungee.getServerInfo( listener.getDefaultServer() );
                         } else
                         {
-                            server = AbstractReconnectHandler.getForcedHost( InitialHandler.this );
-                        }
-                        if ( server == null )
-                        {
-                            server = bungee.getServerInfo( listener.getDefaultServer() );
+                            if ( bungee.getReconnectHandler() != null )
+                            {
+                                server = bungee.getReconnectHandler().getServer( userCon );
+                            } else
+                            {
+                                server = bungee.getServerInfo( listener.getDefaultServer() );
+                            }
                         }
 
                         userCon.connect( server, null, true, ServerConnectEvent.Reason.JOIN_PROXY );
-- 
2.22.0.windows.1

