From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <ivan@mrivanplays.com>
Date: Tue, 16 Jun 2020 10:54:03 +0300
Subject: [PATCH] Configurable brand


diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
index 5e0af575712120c46a06f9f393fd4566f99c575b..e5c113f5616c431affc596ea4c0db64c633c1760 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
@@ -320,5 +320,12 @@ public interface ProxyConfig
      * @return whether to populate query with plugins
      */
     boolean shouldPopulatePluginsWhenQuery();
+
+    /**
+     * Returns the brand proxy will send to the client.
+     *
+     * @return brand
+     */
+    String getBrand();
     // IvanCord end
 }
diff --git a/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java b/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
index 86dc461793e25b62999665b2ee881b07293ad5ae..8c91732f0701029dab1408f5b884e91cdfa8fd79 100644
--- a/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
+++ b/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
@@ -16,6 +16,7 @@ public class IvanCordConfiguration extends WaterfallConfiguration
 
     private boolean kickToLobby = true;
     private String reconnectHandlerType = ReconnectHandlerType.YAML.name();
+    private String brand = "IvanCord by MrIvanPlays";
 
     @Accessors(fluent = true)
     private boolean isTCPFastOpenEnabled = true;
@@ -38,6 +39,7 @@ public class IvanCordConfiguration extends WaterfallConfiguration
         shouldCheckForGlobalOpIndependently = config.getBoolean( "should_check_for_global_op_independently", shouldCheckForGlobalOpIndependently );
         reconnectHandlerType = setupReconnectHandlerType( config.getString( "reconnect_handler_type", reconnectHandlerType ) );
         shouldPopulatePluginsWhenQuery = config.getBoolean( "should_populate_plugins_when_query", shouldPopulatePluginsWhenQuery );
+        brand = config.getString( "brand", brand );
     }
 
     private int setupTFO(int incoming)
diff --git a/proxy/src/main/java/net/md_5/bungee/ServerConnector.java b/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
index e3bf67f76aaba70d7cccaf4dce7c4499cfdf4fa1..28c95e059d9dfab752c5b3bd91b866bb4f486593 100644
--- a/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
+++ b/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
@@ -285,7 +285,7 @@ public class ServerConnector extends PacketHandler
             } else
             {
                 ByteBuf brand = ByteBufAllocator.DEFAULT.heapBuffer();
-                DefinedPacket.writeString( bungee.getName() + " (" + bungee.getVersion() + ")", brand );
+                DefinedPacket.writeString( bungee.getConfig().getBrand(), brand ); // IvanCord - configurable brand
                 user.unsafe().sendPacket( new PluginMessage( user.getPendingConnection().getVersion() >= ProtocolConstants.MINECRAFT_1_13 ? "minecraft:brand" : "MC|Brand", brand, handshakeHandler.isServerForge() ) );
                 brand.release();
             }
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index 9f4a18cd5a94963a58dccfbc79f6fefe702da17f..cbaaccac3c56168eae2cf2272fb8ff05acf80081 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -311,7 +311,7 @@ public class DownstreamBridge extends PacketHandler
             Preconditions.checkState( !serverBrand.contains( bungee.getName() ), "Cannot connect proxy to itself!" );
 
             brand = ByteBufAllocator.DEFAULT.heapBuffer();
-            DefinedPacket.writeString( bungee.getName() + " <- " + serverBrand, brand ); // Waterfall
+            DefinedPacket.writeString( bungee.getConfig().getBrand(), brand ); // Waterfall // IvanCord - custom brand
             pluginMessage.setData( brand );
             brand.release();
             // changes in the packet are ignored so we need to send it manually
