From 538350ee55d4dc2f9c35ceecce63f24147d85117 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Sun, 4 Aug 2019 11:36:40 +0300
Subject: [PATCH] Add option to listen to packets


diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyServer.java b/api/src/main/java/net/md_5/bungee/api/ProxyServer.java
index 1eb8a79b..f6d3524c 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyServer.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyServer.java
@@ -6,6 +6,7 @@ import java.net.InetSocketAddress;
 import java.util.Collection;
 import java.util.Map;
 import java.util.UUID;
+import java.util.function.BiConsumer; // IvanCord
 import java.util.logging.Logger;
 import lombok.Getter;
 import net.md_5.bungee.api.chat.BaseComponent;
@@ -15,9 +16,11 @@ import net.md_5.bungee.api.connection.ProxiedPlayer;
 import net.md_5.bungee.api.plugin.Plugin;
 import net.md_5.bungee.api.plugin.PluginManager;
 import net.md_5.bungee.api.scheduler.TaskScheduler;
+import net.md_5.bungee.protocol.DefinedPacket; // IvanCord
 
 import com.mrivanplays.ivancord.api.plugin.ServicesManager; // IvanCord
 import com.mrivanplays.ivancord.api.command.IvanCordCommandManager; // IvanCord
+import com.mrivanplays.ivancord.api.PacketDirection; // IvanCord
 
 public abstract class ProxyServer
 {
@@ -334,6 +337,22 @@ public abstract class ProxyServer
      */
     public abstract IvanCordCommandManager getIvanCordCommandManager();
 
+    /**
+     * Registers a inbound packet listener.
+     *
+     * @param plugin the plugin assigned to
+     * @param packetListener the inbound packet listener
+     */
+    public abstract void registerInboundPacketListener(Plugin plugin, BiConsumer<DefinedPacket, PacketDirection> packetListener);
+
+    /**
+     * Registers a outbound packet listener
+     *
+     * @param plugin the plugin assigned to
+     * @param packetListener the outbound packet listener
+     */
+    public abstract void registerOutboundPacketListener(Plugin plugin, BiConsumer<DefinedPacket, PacketDirection> packetListener);
+
     // IvanCord end
 
 }
diff --git a/protocol/src/main/java/com/mrivanplays/ivancord/api/PacketDirection.java b/protocol/src/main/java/com/mrivanplays/ivancord/api/PacketDirection.java
new file mode 100644
index 00000000..4fe19ff5
--- /dev/null
+++ b/protocol/src/main/java/com/mrivanplays/ivancord/api/PacketDirection.java
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2019 Ivan Pekov (MrIvanPlays)
+ * Copyright 2019 contributors
+
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in the
+ * Software without restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
+ * and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+
+ * The above copyright notice and this permission notice shall be included in all copies
+ * or substantial portions of the Software.
+
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ **/
+package com.mrivanplays.ivancord.api;
+
+/**
+ * Represents a sent packet direction
+ */
+public enum PacketDirection
+{
+    /**
+     * The packet is being sent to client
+     */
+    TO_CLIENT,
+
+    /**
+     * The packet is being sent to server
+     */
+    TO_SERVER
+}
diff --git a/protocol/src/main/java/com/mrivanplays/ivancord/protocol/PacketListener.java b/protocol/src/main/java/com/mrivanplays/ivancord/protocol/PacketListener.java
new file mode 100644
index 00000000..b7a7fe89
--- /dev/null
+++ b/protocol/src/main/java/com/mrivanplays/ivancord/protocol/PacketListener.java
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2019 Ivan Pekov (MrIvanPlays)
+ * Copyright 2019 contributors
+
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in the
+ * Software without restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies of the Software,
+ * and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+
+ * The above copyright notice and this permission notice shall be included in all copies
+ * or substantial portions of the Software.
+
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ **/
+package com.mrivanplays.ivancord.protocol;
+
+import com.mrivanplays.ivancord.api.PacketDirection;
+import net.md_5.bungee.protocol.DefinedPacket;
+
+public interface PacketListener
+{
+
+    void handle(DefinedPacket packet, PacketDirection direction);
+}
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java b/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java
index a46bbc78..d819ee99 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java
@@ -1,10 +1,15 @@
 package net.md_5.bungee.protocol;
 
+// IvanCord start
+import com.mrivanplays.ivancord.protocol.PacketListener;
+import com.mrivanplays.ivancord.api.PacketDirection;
+// IvanCord end
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufUtil;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.handler.codec.DecoderException;
 import io.netty.handler.codec.MessageToMessageDecoder;
+import java.util.Collection; // IvanCord
 import java.util.List;
 import lombok.AllArgsConstructor;
 import lombok.Setter;
@@ -21,17 +26,21 @@ public class MinecraftDecoder extends MessageToMessageDecoder<ByteBuf>
     @Setter
     private boolean supportsForge = false;
     private final boolean allowEmptyPackets; // Waterfall
+    private Collection<PacketListener> inboundPacketListeners; // IvanCord
 
-    public MinecraftDecoder(Protocol protocol, boolean server, int protocolVersion) {
+    public MinecraftDecoder(Protocol protocol, boolean server, int protocolVersion, Collection<PacketListener> inboundPacketListeners) { // IvanCord
         // Waterfall start
-        this(protocol, server, protocolVersion, false);
+        // IvanCord start
+        this(protocol, server, protocolVersion, false, inboundPacketListeners);
     }
-    public MinecraftDecoder(Protocol protocol, boolean server, int protocolVersion, boolean allowEmptyPackets) {
+    public MinecraftDecoder(Protocol protocol, boolean server, int protocolVersion, boolean allowEmptyPackets, Collection<PacketListener> inboundPacketListeners) {
         // Waterfall end
+        // IvanCord end
         this.protocol = protocol;
         this.server = server;
         this.protocolVersion = protocolVersion;
         this.allowEmptyPackets = allowEmptyPackets; // Waterfall
+        this.inboundPacketListeners = inboundPacketListeners; // IvanCord
     }
 
     @Override
@@ -58,6 +67,12 @@ public class MinecraftDecoder extends MessageToMessageDecoder<ByteBuf>
             {
                 packetTypeInfo = packet.getClass();
                 packet.read( in, prot.getDirection(), protocolVersion );
+                // IvanCord start
+                for ( PacketListener packetListener : inboundPacketListeners )
+                {
+                    packetListener.handle( packet, PacketDirection.valueOf( prot.getDirection().name() ) );
+                }
+                // IvanCord end
 
                 if ( in.isReadable() )
                 {
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftEncoder.java b/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftEncoder.java
index d4b03843..15a74ecb 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftEncoder.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftEncoder.java
@@ -1,5 +1,10 @@
 package net.md_5.bungee.protocol;
 
+// IvanCord start
+import java.util.Collection;
+import com.mrivanplays.ivancord.api.PacketDirection;
+import com.mrivanplays.ivancord.protocol.PacketListener;
+// IvanCord end
 import io.netty.buffer.ByteBuf;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.handler.codec.MessageToByteEncoder;
@@ -15,6 +20,7 @@ public class MinecraftEncoder extends MessageToByteEncoder<DefinedPacket>
     private boolean server;
     @Setter
     private int protocolVersion;
+    private Collection<PacketListener> outboundPacketListeners; // IvanCord
 
     @Override
     protected void encode(ChannelHandlerContext ctx, DefinedPacket msg, ByteBuf out) throws Exception
@@ -22,5 +28,11 @@ public class MinecraftEncoder extends MessageToByteEncoder<DefinedPacket>
         Protocol.DirectionData prot = ( server ) ? protocol.TO_CLIENT : protocol.TO_SERVER;
         DefinedPacket.writeVarInt( prot.getId( msg.getClass(), protocolVersion ), out );
         msg.write( out, prot.getDirection(), protocolVersion );
+        // IvanCord start
+        for ( PacketListener packetListener : outboundPacketListeners )
+        {
+            packetListener.handle( msg, server ? PacketDirection.TO_CLIENT : PacketDirection.TO_SERVER );
+        }
+        // IvanCord end
     }
 }
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 63d9b326..88d277a2 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -34,12 +34,14 @@ import java.util.Map;
 import java.util.MissingResourceException;
 import java.util.PropertyResourceBundle;
 import java.util.ResourceBundle;
+import java.util.Set; // IvanCord
 import java.util.Timer;
 import java.util.TimerTask;
 import java.util.UUID;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.locks.ReadWriteLock;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.BiConsumer; // IvanCord
 import java.util.logging.Handler;
 import java.util.logging.Level;
 import java.util.logging.Logger;
@@ -106,6 +108,8 @@ import com.mrivanplays.ivancord.modules.YamlReconnectHandler;
 import com.mrivanplays.ivancord.SimpleServicesManager;
 import com.mrivanplays.ivancord.api.plugin.ServicesManager;
 import com.mrivanplays.ivancord.api.command.IvanCordCommandManager;
+import com.mrivanplays.ivancord.protocol.PacketListener;
+import com.mrivanplays.ivancord.api.PacketDirection;
 
 /**
  * Main BungeeCord proxy class.
@@ -193,6 +197,10 @@ public class BungeeCord extends ProxyServer
     private ServicesManager servicesManager;
     @Getter
     private IvanCordCommandManager ivanCordCommandManager;
+    private Set<PacketListener> inboundPacketListeners = new HashSet<>();
+    private ReadWriteLock inboundLock = new ReentrantReadWriteLock();
+    private Set<PacketListener> outboundPacketListeners = new HashSet<>();
+    private ReadWriteLock outboundLock = new ReentrantReadWriteLock();
     // IvanCord end
 
     {
@@ -811,4 +819,70 @@ public class BungeeCord extends ProxyServer
     {
         return new BungeeTitle();
     }
+
+    // IvanCord start
+    public Collection<PacketListener> getInboundPacketListeners()
+    {
+        inboundLock.writeLock().lock();
+        try
+        {
+            return inboundPacketListeners;
+        } finally
+        {
+            inboundLock.writeLock().unlock();
+        }
+    }
+
+    public Collection<PacketListener> getOutboundPacketListeners()
+    {
+        outboundLock.writeLock().lock();
+        try
+        {
+            return outboundPacketListeners;
+        } finally
+        {
+            outboundLock.writeLock().unlock();
+        }
+    }
+
+    @Override
+    public void registerInboundPacketListener(Plugin plugin, BiConsumer<DefinedPacket, PacketDirection> packetListener)
+    {
+        inboundLock.readLock().lock();
+        try
+        {
+            inboundPacketListeners.add( (packet, direction) ->
+            {
+                if ( !plugin.isEnabled() )
+                {
+                    return;
+                }
+                packetListener.accept( packet, direction );
+            } );
+        } finally
+        {
+            inboundLock.readLock().unlock();
+        }
+    }
+
+    @Override
+    public void registerOutboundPacketListener(Plugin plugin, BiConsumer<DefinedPacket, PacketDirection> packetListener)
+    {
+        outboundLock.readLock().lock();
+        try
+        {
+            outboundPacketListeners.add( (packet, direction) ->
+            {
+                if ( !plugin.isEnabled() )
+                {
+                    return;
+                }
+                packetListener.accept( packet, direction );
+            } );
+        } finally
+        {
+            outboundLock.readLock().unlock();
+        }
+    }
+    // IvanCord end
 }
diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 5a0ef6b7..a18727d3 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -387,8 +387,8 @@ public final class UserConnection implements ProxiedPlayer
                     protected void initChannel(Channel ch) throws Exception
                     {
                         PipelineUtils.BASE.initChannel( ch );
-                        ch.pipeline().addAfter( PipelineUtils.FRAME_DECODER, PipelineUtils.PACKET_DECODER, new MinecraftDecoder( Protocol.HANDSHAKE, false, getPendingConnection().getVersion(), bungee.getConfig().isAllowEmptyPackets() ) ); // Waterfall
-                        ch.pipeline().addAfter( PipelineUtils.FRAME_PREPENDER, PipelineUtils.PACKET_ENCODER, new MinecraftEncoder( Protocol.HANDSHAKE, false, getPendingConnection().getVersion() ) );
+                        ch.pipeline().addAfter( PipelineUtils.FRAME_DECODER, PipelineUtils.PACKET_DECODER, new MinecraftDecoder( Protocol.HANDSHAKE, false, getPendingConnection().getVersion(), bungee.getConfig().isAllowEmptyPackets(), BungeeCord.getInstance().getInboundPacketListeners() ) ); // Waterfall // IvanCord
+                        ch.pipeline().addAfter( PipelineUtils.FRAME_PREPENDER, PipelineUtils.PACKET_ENCODER, new MinecraftEncoder( Protocol.HANDSHAKE, false, getPendingConnection().getVersion(), BungeeCord.getInstance().getOutboundPacketListeners() ) ); // IvanCord
                         ch.pipeline().get( HandlerBoss.class ).setHandler( new ServerConnector( bungee, UserConnection.this, target ) );
                     }
                 };
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/PingHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/PingHandler.java
index 6bc14d67..2db9f157 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/PingHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/PingHandler.java
@@ -34,9 +34,9 @@ public class PingHandler extends PacketHandler
     public void connected(ChannelWrapper channel) throws Exception
     {
         this.channel = channel;
-        MinecraftEncoder encoder = new MinecraftEncoder( Protocol.HANDSHAKE, false, protocol );
+        MinecraftEncoder encoder = new MinecraftEncoder( Protocol.HANDSHAKE, false, protocol, BungeeCord.getInstance().getOutboundPacketListeners() ); // IvanCord
 
-        channel.getHandle().pipeline().addAfter( PipelineUtils.FRAME_DECODER, PipelineUtils.PACKET_DECODER, new MinecraftDecoder( Protocol.STATUS, false, ProxyServer.getInstance().getProtocolVersion() ) );
+        channel.getHandle().pipeline().addAfter( PipelineUtils.FRAME_DECODER, PipelineUtils.PACKET_DECODER, new MinecraftDecoder( Protocol.STATUS, false, ProxyServer.getInstance().getProtocolVersion(), BungeeCord.getInstance().getInboundPacketListeners() ) ); // IvanCord
         channel.getHandle().pipeline().addAfter( PipelineUtils.FRAME_PREPENDER, PipelineUtils.PACKET_ENCODER, encoder );
 
         channel.write( new Handshake( protocol, target.getAddress().getHostString(), target.getAddress().getPort(), 1 ) );
diff --git a/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java b/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
index 39ed623b..77a39bc3 100644
--- a/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
+++ b/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
@@ -69,8 +69,8 @@ public class PipelineUtils
             }
             // IvanCord end
             ch.pipeline().addBefore( FRAME_DECODER, LEGACY_DECODER, new LegacyDecoder() );
-            ch.pipeline().addAfter( FRAME_DECODER, PACKET_DECODER, new MinecraftDecoder( Protocol.HANDSHAKE, true, ProxyServer.getInstance().getProtocolVersion() ) );
-            ch.pipeline().addAfter( FRAME_PREPENDER, PACKET_ENCODER, new MinecraftEncoder( Protocol.HANDSHAKE, true, ProxyServer.getInstance().getProtocolVersion() ) );
+            ch.pipeline().addAfter( FRAME_DECODER, PACKET_DECODER, new MinecraftDecoder( Protocol.HANDSHAKE, true, ProxyServer.getInstance().getProtocolVersion(), BungeeCord.getInstance().getInboundPacketListeners() ) ); // IvanCord
+            ch.pipeline().addAfter( FRAME_PREPENDER, PACKET_ENCODER, new MinecraftEncoder( Protocol.HANDSHAKE, true, ProxyServer.getInstance().getProtocolVersion(), BungeeCord.getInstance().getOutboundPacketListeners() ) ); // IvanCord
             ch.pipeline().addBefore( FRAME_PREPENDER, LEGACY_KICKER, legacyKicker );
             ch.pipeline().get( HandlerBoss.class ).setHandler( new InitialHandler( BungeeCord.getInstance(), listener ) );
 
-- 
2.22.0.windows.1

