From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Sat, 15 Feb 2020 12:43:03 +0200
Subject: [PATCH] Close connection if invalid handshake packet state was
 received instead of throwing exception


diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index ed525efadd950d34a0f6638117facedc81b76143..3df4d910c607193d2a83cbc4c5732ae13c2e52ce 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -354,7 +354,11 @@ public class InitialHandler extends PacketHandler implements PendingConnection
                 }
                 break;
             default:
+                /* // IvanCord start - close connection if invalid handshake packet state was received instead of throwing exception
                 throw new QuietException( "Cannot request protocol " + handshake.getRequestedProtocol() );
+                 */
+                ch.close();
+                // IvanCord end
         }
     }
 
