From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <ivan@mrivanplays.com>
Date: Tue, 16 Jun 2020 10:54:03 +0300
Subject: [PATCH] Configurable brand


diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
index 3546e3eb40e0a066d129a58cc1af607e9a265fcf..f56db5737913afc343b332fdfffc7c6adce8e87f 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
@@ -294,5 +294,12 @@ public interface ProxyConfig
      */
     String getReconnectHandlerType();
 
+    /**
+     * Returns the brand proxy will send to the client.
+     *
+     * @return brand
+     */
+    String getBrand();
+
     // IvanCord end
 }
diff --git a/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java b/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
index 1111e912f1668b5cb330c13fd27972d5875b31b8..161f0bf5b4fa0e5fefdbebfdea178be2252116a0 100644
--- a/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
+++ b/proxy/src/main/java/com/mrivanplays/ivancord/conf/IvanCordConfiguration.java
@@ -6,6 +6,7 @@ import com.mrivanplays.ivancord.ReconnectHandlerType;
 import io.github.waterfallmc.waterfall.conf.WaterfallConfiguration;
 import lombok.Getter;
 import lombok.experimental.Accessors;
+import net.md_5.bungee.BungeeCord;
 import net.md_5.bungee.api.ProxyServer;
 import net.md_5.bungee.conf.YamlConfig;
 import com.mrivanplays.ivancord.tcpfastopen.TCPFastOpenMode;
@@ -17,6 +18,7 @@ public class IvanCordConfiguration extends WaterfallConfiguration
     private boolean kickToLobby = true;
     private boolean disableTabListRewrite = isOnlineMode();
     private String reconnectHandlerType = ReconnectHandlerType.YAML.name();
+    private String brand = "IvanCord by MrIvanPlays";
 
     @Accessors(fluent = true)
     private boolean isTCPFastOpenEnabled = true;
@@ -34,6 +36,7 @@ public class IvanCordConfiguration extends WaterfallConfiguration
         isTCPFastOpenEnabled = config.getBoolean( "tcp_fast_open_enabled", isTCPFastOpenEnabled );
         getTCPFastOpenMode = setupTFO( config.getInt( "tcp_fast_open_mode", getTCPFastOpenMode ) );
         reconnectHandlerType = setupReconnectHandlerType( config.getString( "reconnect_handler_type", reconnectHandlerType ) );
+        brand = config.getString( "brand", brand );
     }
 
     private int setupTFO(int incoming)
diff --git a/proxy/src/main/java/net/md_5/bungee/ServerConnector.java b/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
index cf016d3e0384e8a39bb258bb4a0864eb01cd3747..a6e052baee4a862b1db94869b6a92e385324accb 100644
--- a/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
+++ b/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
@@ -273,7 +273,7 @@ public class ServerConnector extends PacketHandler
             } else
             {
                 ByteBuf brand = ByteBufAllocator.DEFAULT.heapBuffer();
-                DefinedPacket.writeString( bungee.getName() + " (" + bungee.getVersion() + ")", brand );
+                DefinedPacket.writeString( bungee.getConfig().getBrand(), brand ); // IvanCord - configurable brand
                 user.unsafe().sendPacket( new PluginMessage( user.getPendingConnection().getVersion() >= ProtocolConstants.MINECRAFT_1_13 ? "minecraft:brand" : "MC|Brand", brand, handshakeHandler.isServerForge() ) );
                 brand.release();
             }
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index b138a8ec499276184e55cef57a9db08ebe57d9aa..300f71ebe0684c1b7fa8c61838adc5f95c685861 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -314,7 +314,7 @@ public class DownstreamBridge extends PacketHandler
             Preconditions.checkState( !serverBrand.contains( bungee.getName() ), "Cannot connect proxy to itself!" );
 
             brand = ByteBufAllocator.DEFAULT.heapBuffer();
-            DefinedPacket.writeString( bungee.getName() + " <- " + serverBrand, brand ); // Waterfall
+            DefinedPacket.writeString( bungee.getConfig().getBrand(), brand ); // Waterfall // IvanCord - custom brand
             pluginMessage.setData( brand );
             brand.release();
             // changes in the packet are ignored so we need to send it manually
