From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Antoine L <antoine.l@erisium.com>
Date: Tue, 15 Jun 2021 18:21:03 +0300
Subject: [PATCH] Use the new ActionBar packet to send action bars for 1.17+
 clients


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/AbstractPacketHandler.java b/protocol/src/main/java/net/md_5/bungee/protocol/AbstractPacketHandler.java
index 534bcc41aeb08b630c7f816ba7038f27d2c2419d..179a06aca2b3f06e16ecb40663a632517b9ba6a1 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/AbstractPacketHandler.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/AbstractPacketHandler.java
@@ -1,5 +1,6 @@
 package net.md_5.bungee.protocol;
 
+import net.md_5.bungee.protocol.packet.ActionBar; // IvanCord
 import net.md_5.bungee.protocol.packet.BossBar;
 import net.md_5.bungee.protocol.packet.Chat;
 import net.md_5.bungee.protocol.packet.ClearTitles;
@@ -247,5 +248,9 @@ public abstract class AbstractPacketHandler
     public void handle(EntityAttributes entityAttributes) throws Exception
     {
     }
+
+    public void handle(ActionBar actionBar) throws Exception
+    {
+    }
     // IvanCord end
 }
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
index 5e347382e8830bcf47f6ba2e4e5d2041c4c7ccf7..d28c32a286ef64b0106c89d47f03fe7bd89a3729 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
@@ -11,6 +11,7 @@ import gnu.trove.map.hash.TObjectIntHashMap;
 import java.util.function.Supplier;
 import lombok.Data;
 import lombok.Getter;
+import net.md_5.bungee.protocol.packet.ActionBar;
 import net.md_5.bungee.protocol.packet.BossBar;
 import net.md_5.bungee.protocol.packet.Chat;
 import net.md_5.bungee.protocol.packet.ClearTitles;
@@ -430,6 +431,17 @@ public enum Protocol
                     map( ProtocolConstants.MINECRAFT_1_19_3, 0x36 ),
                     map( ProtocolConstants.MINECRAFT_1_19_4, 0x3A )
             );
+            // IvanCord start
+            TO_CLIENT.registerPacket(
+                    ActionBar.class,
+                    ActionBar::new,
+                    map( ProtocolConstants.MINECRAFT_1_17, 0x41 ),
+                    map( ProtocolConstants.MINECRAFT_1_19, 0x40 ),
+                    map( ProtocolConstants.MINECRAFT_1_19_1, 0x43 ),
+                    map( ProtocolConstants.MINECRAFT_1_19_3, 0x42 ),
+                    map( ProtocolConstants.MINECRAFT_1_19_4, 0x46 )
+            );
+            // IvanCord end
 
             TO_SERVER.registerPacket(
                     KeepAlive.class,
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/packet/ActionBar.java b/protocol/src/main/java/net/md_5/bungee/protocol/packet/ActionBar.java
new file mode 100644
index 0000000000000000000000000000000000000000..c553ba02d7f13c83f72f5e016bda828141630d97
--- /dev/null
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/packet/ActionBar.java
@@ -0,0 +1,39 @@
+// IvanCord start
+package net.md_5.bungee.protocol.packet;
+
+import io.netty.buffer.ByteBuf;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.NoArgsConstructor;
+import net.md_5.bungee.protocol.AbstractPacketHandler;
+import net.md_5.bungee.protocol.DefinedPacket;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@EqualsAndHashCode(callSuper = false)
+public class ActionBar extends DefinedPacket
+{
+
+    private String text;
+
+    @Override
+    public void read(ByteBuf buf)
+    {
+        text = readString( buf, 262144 );
+    }
+
+    @Override
+    public void write(ByteBuf buf)
+    {
+        writeString( text, buf );
+    }
+
+    @Override
+    public void handle(AbstractPacketHandler handler) throws Exception
+    {
+        handler.handle( this );
+    }
+}
+// IvanCord end
diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index d89e9bebb9c30d6bbab443226aafbdc5e1a6bf08..d8a96e52c7da233f1fec3ad80a0e0a307b731154 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -555,13 +555,17 @@ public final class UserConnection implements ProxiedPlayer
         // transform score components
         message = ChatComponentTransformer.getInstance().transform( this, true, message );
 
-        if ( position == ChatMessageType.ACTION_BAR && getPendingConnection().getVersion() < ProtocolConstants.MINECRAFT_1_17 )
+        if ( position == ChatMessageType.ACTION_BAR ) // IvanCord - remove check
         {
             // Versions older than 1.11 cannot send the Action bar with the new JSON formattings
             // Fix by converting to a legacy message, see https://bugs.mojang.com/browse/MC-119145
             if ( getPendingConnection().getVersion() <= ProtocolConstants.MINECRAFT_1_10 )
             {
                 sendMessage( position, sender, ComponentSerializer.toString( new TextComponent( BaseComponent.toLegacyText( message ) ) ) );
+            } else if ( getPendingConnection().getVersion() >= ProtocolConstants.MINECRAFT_1_17 ) // IvanCord start
+            {
+                unsafe.sendPacket( new net.md_5.bungee.protocol.packet.ActionBar( ComponentSerializer.toString( message ) ) );
+                // IvanCord end
             } else
             {
                 net.md_5.bungee.protocol.packet.Title title = new net.md_5.bungee.protocol.packet.Title();
