From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivan Pekov <ivan@mrivanplays.com>
Date: Tue, 22 Sep 2020 16:09:27 +0300
Subject: [PATCH] Add some netty options from Velocity


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
index 0a19f6d7346f282e98bd4d5ac781bd6c5ef0be95..0b1751e5879c8ec0e2dc6c930188a6ff84167d7e 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
@@ -39,6 +39,15 @@ public class BungeeCordLauncher
         }
         // Waterfall end
 
+        // IvanCord start - resource leak detector changes from VelocityPowered/Velocity
+        // Disable the resource leak detector by default as it reduces performance. Allow the user to
+        // override this if desired.
+        if ( System.getProperty( "io.netty.leakDetection.level" ) == null )
+        {
+            io.netty.util.ResourceLeakDetector.setLevel( io.netty.util.ResourceLeakDetector.Level.DISABLED );
+        }
+        // IvanCord end
+
         OptionParser parser = new OptionParser();
         parser.allowsUnrecognizedOptions();
         parser.acceptsAll( Arrays.asList( "help" ), "Show the help" );
