From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gerrygames <gecam59@gmail.com>
Date: Tue, 13 Oct 2020 18:21:24 +0300
Subject: [PATCH] Fix infinite relay message accumulation

Co-authored-by: Ivan Pekov <ivan@mrivanplays.com>

diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index fb6c7282ffe16f700fe2ab2ea714f92335dcd72e..830504439884cb4901666c417420d3df999ba969 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -164,7 +164,7 @@ public class InitialHandler extends PacketHandler implements PendingConnection
     public void handle(PluginMessage pluginMessage) throws Exception
     {
         // TODO: Unregister?
-        if ( PluginMessage.SHOULD_RELAY.apply( pluginMessage ) )
+        if ( PluginMessage.SHOULD_RELAY.apply( pluginMessage ) && !relayMessages.contains( pluginMessage ) ) // IvanCord
         {
             relayMessages.add( pluginMessage );
         }
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
index 9708400babb47859fbee4fcc65988c27524bda1b..fb373250a6f8a46adf9b207ba92f7a7fe0d1a83d 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
@@ -374,7 +374,7 @@ public class UpstreamBridge extends PacketHandler
         }
 
         // TODO: Unregister as well?
-        if ( PluginMessage.SHOULD_RELAY.apply( pluginMessage ) )
+        if ( PluginMessage.SHOULD_RELAY.apply( pluginMessage ) && !con.getPendingConnection().getRelayMessages().contains( pluginMessage ) ) // IvanCord
         {
             con.getPendingConnection().getRelayMessages().add( pluginMessage );
         }
