From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gerrygames <gecam59@gmail.com>
Date: Tue, 13 Oct 2020 18:21:24 +0300
Subject: [PATCH] Fix infinite relay message accumulation

Co-authored-by: Ivan Pekov <ivan@mrivanplays.com>

diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index 6112ff552a789e3e2cdc68c3083b9c4919037e7c..3a17cce5cd86437467ea188b9ebbc3b1fb1cb3c6 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -165,7 +165,7 @@ public class InitialHandler extends PacketHandler implements PendingConnection
     public void handle(PluginMessage pluginMessage) throws Exception
     {
         // TODO: Unregister?
-        if ( PluginMessage.SHOULD_RELAY.apply( pluginMessage ) )
+        if ( PluginMessage.SHOULD_RELAY.apply( pluginMessage ) && !relayMessages.contains( pluginMessage ) ) // IvanCord
         {
             relayMessages.add( pluginMessage );
         }
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
index b5f95ae45eaf92693a196be0fbc669372fca4c84..6311ef0a8440c8b7a2b7d02f4826078627b0b574 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
@@ -376,7 +376,7 @@ public class UpstreamBridge extends PacketHandler
         }
 
         // TODO: Unregister as well?
-        if ( PluginMessage.SHOULD_RELAY.apply( pluginMessage ) )
+        if ( PluginMessage.SHOULD_RELAY.apply( pluginMessage ) && !con.getPendingConnection().getRelayMessages().contains( pluginMessage ) ) // IvanCord
         {
             con.getPendingConnection().getRelayMessages().add( pluginMessage );
         }
