From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivan Pekov <ivan@mrivanplays.com>
Date: Mon, 15 Nov 2021 08:05:06 +0200
Subject: [PATCH] Add isLegacy to ProxyPingEvent


diff --git a/api/src/main/java/net/md_5/bungee/api/event/ProxyPingEvent.java b/api/src/main/java/net/md_5/bungee/api/event/ProxyPingEvent.java
index ddee993a41a56425897e50c3664341e932ef7a64..562b1f34a8a0a0cf7f5a84f78bf0bfd8a2952f05 100644
--- a/api/src/main/java/net/md_5/bungee/api/event/ProxyPingEvent.java
+++ b/api/src/main/java/net/md_5/bungee/api/event/ProxyPingEvent.java
@@ -25,10 +25,25 @@ public class ProxyPingEvent extends AsyncEvent<ProxyPingEvent>
      */
     private ServerPing response;
 
+    // IvanCord start
+    /**
+     * Whether this event was called for a legacy ping
+     */
+    private final boolean legacy;
+    // IvanCord end
+
     public ProxyPingEvent(PendingConnection connection, ServerPing response, Callback<ProxyPingEvent> done)
+    // IvanCord start
+    {
+        this( connection, response, false, done );
+    }
+
+    public ProxyPingEvent(PendingConnection connection, ServerPing response, boolean legacy, Callback<ProxyPingEvent> done)
+    // IvanCord end
     {
         super( done );
         this.connection = connection;
         this.response = response;
+        this.legacy = legacy; // IvanCord
     }
 }
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index e5771ca34e95bb07f9241a4ce7cdf7ebd1e342c7..1617d8202fc362f8c50656606a08c3d675ff1af8 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -243,7 +243,7 @@ public class InitialHandler extends PacketHandler implements PendingConnection
                     }
                 };
 
-                bungee.getPluginManager().callEvent( new ProxyPingEvent( InitialHandler.this, result, callback ) );
+                bungee.getPluginManager().callEvent( new ProxyPingEvent( InitialHandler.this, result, legacy, callback ) ); // IvanCord
             }
         };
 
