From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivan Pekov <ivan@mrivanplays.com>
Date: Tue, 22 Sep 2020 16:09:27 +0300
Subject: [PATCH] Add some netty options from Velocity


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
index 55a0c815611329b62cee1837814a27b31f80ce0d..e55895d65616be8116b148cd804b1d08e3c2abc6 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
@@ -34,6 +34,15 @@ public class BungeeCordLauncher
         }
         // Waterfall end
 
+        // IvanCord start - resource leak detector changes from VelocityPowered/Velocity
+        // Disable the resource leak detector by default as it reduces performance. Allow the user to
+        // override this if desired.
+        if ( System.getProperty( "io.netty.leakDetection.level" ) == null )
+        {
+            io.netty.util.ResourceLeakDetector.setLevel( io.netty.util.ResourceLeakDetector.Level.DISABLED );
+        }
+        // IvanCord end
+
         OptionParser parser = new OptionParser();
         parser.allowsUnrecognizedOptions();
         parser.acceptsAll( Arrays.asList( "help" ), "Show the help" );
