From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gerrygames <gecam59@gmail.com>
Date: Tue, 13 Oct 2020 18:21:24 +0300
Subject: [PATCH] Fix infinite relay message accumulation

Co-authored-by: Ivan Pekov <ivan@mrivanplays.com>

diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index 308e94db6233d593a60f0e4c23f75549db42ee54..ce4202d96eee316b8fd1061c8885d30ed71f35f0 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -164,7 +164,7 @@ public class InitialHandler extends PacketHandler implements PendingConnection
     public void handle(PluginMessage pluginMessage) throws Exception
     {
         // TODO: Unregister?
-        if ( PluginMessage.SHOULD_RELAY.apply( pluginMessage ) )
+        if ( PluginMessage.SHOULD_RELAY.apply( pluginMessage ) && !relayMessages.contains( pluginMessage ) ) // IvanCord
         {
             relayMessages.add( pluginMessage );
         }
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
index 4603408e1af16a2d8e71b08dbc294feb7cc8302a..4e8644e05f81dcd268ea245b11b3cd40d0455f87 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
@@ -364,7 +364,7 @@ public class UpstreamBridge extends PacketHandler
         }
 
         // TODO: Unregister as well?
-        if ( PluginMessage.SHOULD_RELAY.apply( pluginMessage ) )
+        if ( PluginMessage.SHOULD_RELAY.apply( pluginMessage ) && !con.getPendingConnection().getRelayMessages().contains( pluginMessage ) ) // IvanCord
         {
             con.getPendingConnection().getRelayMessages().add( pluginMessage );
         }
