From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivan Pekov <ivan@mrivanplays.com>
Date: Tue, 22 Sep 2020 16:09:27 +0300
Subject: [PATCH] Add some netty options from Velocity


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
index dc32c863578c9d06f1e08368e109ec3af14fc7fb..6f7386bab6bbc26464244eec79044924358a18a4 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
@@ -16,6 +16,31 @@ import net.md_5.bungee.command.ConsoleCommandSender;
 public class BungeeCordLauncher
 {
 
+    // IvanCord start - velocity's netty stuff
+    // Credits: VelocityPowered/Velocity
+    static
+    {
+        // By default, Netty allocates 16MiB arenas for the PooledByteBufAllocator. This is too much
+        // memory for Minecraft, which imposes a maximum packet size of 2MiB! We'll use 4MiB as a more
+        // sane default.
+        //
+        // Note: io.netty.allocator.pageSize << io.netty.allocator.maxOrder is the formula used to
+        // compute the chunk size. We lower maxOrder from its default of 11 to 9. (We also use a null
+        // check, so that the user is free to choose another setting if need be.)
+        if ( System.getProperty( "io.netty.allocator.maxOrder" ) == null )
+        {
+            System.setProperty( "io.netty.allocator.maxOrder", "9" );
+        }
+
+        // Disable the resource leak detector by default as it reduces performance. Allow the user to
+        // override this if desired.
+        if ( System.getProperty( "io.netty.leakDetection.level" ) == null )
+        {
+            io.netty.util.ResourceLeakDetector.setLevel( io.netty.util.ResourceLeakDetector.Level.DISABLED );
+        }
+    }
+    // IvanCord end
+
     public static void main(String[] args) throws Exception
     {
         Security.setProperty( "networkaddress.cache.ttl", "30" );
