From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ivan Pekov <ivan@mrivanplays.com>
Date: Wed, 13 Jan 2021 19:53:00 +0200
Subject: [PATCH] fixup! Brigadier command api


diff --git a/api/src/main/java/com/mrivanplays/ivancord/api/command/CommandDispatcherProxy.java b/api/src/main/java/com/mrivanplays/ivancord/api/command/CommandDispatcherProxy.java
index 6b954ea5f9773be06796995227cb66285a4f72ea..cf56ab1b54e8c5a4d56ccd54824985e60511663b 100644
--- a/api/src/main/java/com/mrivanplays/ivancord/api/command/CommandDispatcherProxy.java
+++ b/api/src/main/java/com/mrivanplays/ivancord/api/command/CommandDispatcherProxy.java
@@ -180,7 +180,7 @@ public final class CommandDispatcherProxy
         commandName = commandName.toLowerCase();
         if ( dispatcher.getRoot().getChild( commandName ) != null )
         {
-            return !( sender instanceof ProxiedPlayer ) || !proxy.getDisabledCommands().contains( commandName );
+            return !( sender instanceof ProxiedPlayer ) || proxy.getDisabledCommands() == null ||  !proxy.getDisabledCommands().contains( commandName );
         }
         return false;
     }
