From 781b386d40db55943cc1c0886f08f4244b9a34b7 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <pekov.ivan@abv.bg>
Date: Sun, 19 May 2019 16:07:09 +0300
Subject: [PATCH] Add player teleport api


diff --git a/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java b/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java
index 2374ff5e..85e42a4a 100644
--- a/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java
+++ b/api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java
@@ -423,4 +423,15 @@ public interface ProxiedPlayer extends Connection, CommandSender
     void setPosition(Position position);
 
     // IvanCord end
+
+    // IvanCord start - add teleport api
+
+    /**
+     * Teleports the current player to another player
+     *
+     * @param other teleported to
+     */
+    void teleport(ProxiedPlayer other);
+
+    // IvanCord end
 }
diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index dcd2ea2a..24ff580f 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -785,6 +785,23 @@ public final class UserConnection implements ProxiedPlayer
     }
     // IvanCord end
 
+
+    // IvanCord start - add teleport api
+    @Override
+    public void teleport(ProxiedPlayer other)
+    {
+        ServerInfo targetServer = other.getServer().getInfo();
+        if ( !getServer().getInfo().equals( targetServer ) )
+        {
+            connect( targetServer );
+            setPosition( other.getPosition() );
+        } else
+        {
+            setPosition( other.getPosition() );
+        }
+    }
+    // IvanCord end
+
     // Waterfall start
     public boolean isDisableEntityMetadataRewrite() {
         return disableEntityMetadaRewrite;
-- 
2.21.0.windows.1

